<!DOCTYPE html>
<html lang="zh-Hant">
<head>
	<meta charset="utf-8" />
	<meta http-equiv="X-UA-Compatible" content="IE=edge, chrome=1">
	<meta http-equiv="content-type" content="text/html; charset=UTF-8;charset=utf-8">
	<meta name="viewport" content="width=device-width,initial-scale=1, maximum-scale=1">
<meta name="author" content="Liang-Bo Wang" />
<meta name="description" content="æœ¬æ–‡å…§å®¹ä¸»è¦ä¾†è‡ª Barak Cohen æ•™æˆçµ¦çš„æ•¸å ‚èª²çš„ç­†è¨˜ï¼Œä»¥ Systems Biology çš„è§’åº¦ä¾†çœ‹ coding/noncoding variant modeling å’Œç›¸é—œå¯¦é©— MPRAã€‚" />
<meta name="keywords" content="zh, bio">
	<link rel="icon" href="data:image/svg+xml,<svg xmlns=%22http://www.w3.org/2000/svg%22 viewBox=%220 0 100 100%22><text y=%22.95em%22 font-size=%2295%22>ğŸ“–</text></svg>">
	<title>Variantsã€eQTLã€MPRA</title>

	<!-- og -->
<meta property="og:locale" content="zh_TW"/>
<meta property="og:site_name" content="Liang-Bo Wang's Blog"/>
<meta property="og:title" content="Variantsã€eQTLã€MPRA"/>
<meta property="og:description" content="æœ¬æ–‡å…§å®¹ä¸»è¦ä¾†è‡ª Barak Cohen æ•™æˆçµ¦çš„æ•¸å ‚èª²çš„ç­†è¨˜ï¼Œä»¥ Systems Biology çš„è§’åº¦ä¾†çœ‹ coding/noncoding variant modeling å’Œç›¸é—œå¯¦é©— MPRAã€‚"/>
<meta property="og:locale" content=""/>
<meta property="og:url" content="https://blog.liang2.tw/posts/2017/06/variants-eqtl-mpra/"/>
<meta property="og:type" content="article"/>
<meta property="article:published_time" content="2017-06-20 00:00:00-05:00"/>
<meta property="article:modified_time" content="2022-02-20 20:07:45.295570-06:00"/>
<meta property="article:author" content="https://blog.liang2.tw/author/liang-bo-wang.html">
<meta property="article:section" content="Bioinfo"/>
<meta property="article:tag" content="zh"/>
<meta property="article:tag" content="bio"/>
<meta name="twitter:card" content="summary">
<meta name="twitter:site" content="@ccwang002">
<meta name="twitter:title" content="Variantsã€eQTLã€MPRA">
<meta name="twitter:description" content="æœ¬æ–‡å…§å®¹ä¸»è¦ä¾†è‡ª Barak Cohen æ•™æˆçµ¦çš„æ•¸å ‚èª²çš„ç­†è¨˜ï¼Œä»¥ Systems Biology çš„è§’åº¦ä¾†çœ‹ coding/noncoding variant modeling å’Œç›¸é—œå¯¦é©— MPRAã€‚">

	<!-- Feeds -->
	<link href="https://blog.liang2.tw/feeds/all.atom.xml"
		type="application/atom+xml" rel="alternate" title="Liang-Bo Wang's Blog" />

	<link rel="stylesheet" type="text/css" href="https://blog.liang2.tw/theme/styles/styles.css" />
	<link rel="stylesheet" type="text/css" href="https://blog.liang2.tw/theme/katex-0.15.3/katex.min.css">
</head>

<body class="post-template">
	<header class="site-head">
		<h1 class="blog-title"><a href="https://blog.liang2.tw">Liang-Bo Wang's Blog</a></h1>
		<h2 class="blog-description">
			<a href="https://blog.liang2.tw/about-me/">About</a> |
			<a href="https://blog.liang2.tw/talks/">Talks</a> |
			<a href="https://blog.liang2.tw/archives.html">Archives</a> |
			<button class="theme-toggle" aria-label="auto" aria-live="polite">ğŸŒ™</button>
		</h2>
	</header>
	<main id="content" class="content" role="main">
<article class="post">
    <span class="post-meta">
        <time datetime="2017-06-20">Jun 20, 2017</time>
        in <a href="https://blog.liang2.tw/category/bioinfo/">Bioinfo</a>
    </span>
    <h1 class="post-title">
        <a href="https://blog.liang2.tw/posts/2017/06/variants-eqtl-mpra/" rel="bookmark" title="Permalink to Variantsã€eQTLã€MPRA">Variantsã€eQTLã€MPRA</a>
    </h1>
    <span class="post-meta">
                <a href="https://blog.liang2.tw/tag/zh/">zh</a>
                <a href="https://blog.liang2.tw/tag/bio/">bio</a>
    </span>
    <section class="post-content"><p>Computational Biology å’Œ Bioinformatics åœ¨ç¾åœ¨å¯èƒ½å€åˆ†ä¸å¤§ï¼Œæœ¬æ–‡ä¹Ÿä¸æ‰“ç®—æ·±ç©¶å…©å®šç¾©ï¼Œä½†ä»–å€‘å¤§è‡´èƒ½ä»£è¡¨å…©å¤§é¡å°‡é›»è…¦ç§‘å­¸ã€ç¨‹å¼é‹ç”¨åœ¨ç”Ÿç‰©ä¸Šçš„ç ”ç©¶ã€‚</p>
<p>åœ¨ç¢©ç­ï¼Œæˆ‘çš„å¯¦é©—å®¤ä¸€ç›´é¼“å‹µæˆ‘å€‘å»æƒ³æ–°çš„æ¼”ç®—æ³•ï¼ŒæŠŠæŸç¨®é æ¸¬åšå¾—æ›´å¥½æˆ–è€…å¿«ï¼Œæˆ–è€…é‹ç”¨æ›´å¤šä¾†æºçš„æ•¸æ“šï¼›åšæ–°çš„å·¥å…·ï¼›æ•´åˆå‡ºæ–°çš„è³‡æ–™åº«ã€‚é€™äº›æ‡‰ç”¨éƒ½æœ‰ä»–å€‘çš„ç ”ç©¶åƒ¹å€¼ï¼Œä¹Ÿéœ€è¦å¤§é‡çš„æŠ€è¡“æŠ•å…¥ï¼Œå³ä¾¿åœ¨ç™¼è¡¨ä¸Šä¸¦ä¸æœƒæ”¾å…¥é€™äº›ç´°ç¯€ã€‚é€™é¡ç ”ç©¶æ¯”è¼ƒåå‘ Bioinformaticsã€‚</p>
<p>ä¾† WashU å‰ï¼Œæˆ‘æœŸè¨±è‡ªå·±ç¹¼çºŒå¾€ Bioinformatics æ·±å…¥ã€‚ç„¶è€Œï¼Œåœ¨éå»çš„æ•¸æœˆè£¡ï¼Œå³ä¾¿æˆ‘ä»æŠ•å…¥åœ¨é€™äº›æ•¸æ“šåˆ†æèˆ‡å·¥å…·é–‹ç™¼ä¸Šï¼Œå¦ä¸€å¤§éƒ¨ä»½çš„æ™‚é–“ï¼Œæˆ‘ç¶“æ­·äº†è¨±å¤šé—œæ–¼æ¨¡å‹ï¼Œæˆ–è€…ï¼Œé—œæ–¼ã€Œå¦‚ä½•å›ç­”é‡è¦çš„ç”Ÿç‰©å•é¡Œã€ä¸Šçš„è¨è«–ï¼Œæœ‰äº†è¼ƒç¢©ç­è¨“ç·´ä¸åŒçš„å•Ÿç™¼ã€‚é€™å¦ä¸€é¡ç ”ç©¶æ¯”è¼ƒåå‘ Computational Biologyã€‚</p>
<p>æœ¬æ–‡æƒ³ç”¨å¦ä¸€å€‹è§’åº¦ä¾†çœ‹æ‰€è¬‚çš„ã€Œmodelingã€ã€‚å…§å®¹ä¸»è¦ä¾†è‡ª <a href="http://genetics.wustl.edu/bclab/">Barak Cohen</a> æ•™æˆçµ¦çš„æ•¸å ‚èª²çš„ç­†è¨˜ï¼Œä¸»é¡Œç‚º <em>Coding and Noncoding Variant</em>ã€‚æˆ‘ç”Ÿç‰©èƒŒæ™¯ä¸è¶³ï¼Œå¦‚æœç­†è¨˜æœ‰ä»»ä½•éŒ¯èª¤ï¼Œç…©è«‹å‘ŠçŸ¥ã€‚</p>
<p><strong>Conflict of Interest</strong>: Cohen Lab é–‹ç™¼äº† <a href="http://www.pnas.org/content/109/47/19498.short">CRE-seq</a> (<em>cis</em>-regulatory element by sequencing)ï¼Œå…¶ä¸­ä¸€ç¨® MPRA (Massively Parallel Reporter Assay) æŠ€è¡“ã€‚</p>
<div class="toc">
<ul>
<li><a href="#coding-vs-noncoding-variants">Coding vs noncoding variants</a><ul>
<li><a href="#noncoding-elements">Noncoding elements</a></li>
<li><a href="#noncoding-variants">Noncoding variants</a></li>
<li><a href="#endophenotypes">Endophenotypes</a></li>
</ul>
</li>
<li><a href="#eqtl">eQTL</a></li>
<li><a href="#mpra">MPRA</a></li>
<li><a href="#conclusion">Conclusion</a></li>
</ul>
</div>
<h2 id="coding-vs-noncoding-variants">Coding vs noncoding variants</h2>
<p>é¦–å…ˆä¾†è«‡è«‡ coding å’Œ noncoding variantã€‚èª²å ‚ä¸Šè€å¸«è®“æˆ‘å€‘è‡ªç”±è¾¯è«–ç ”ç©¶å…©è€…çš„ã€Œå„ªç¼ºé»ã€ï¼Œäº¦æˆ–ï¼Œå¦‚æœä½ æ˜¯ PI æ¯”è¼ƒæƒ³ç ”ç©¶å“ªå€‹ï¼Œé¢è‡¨çš„å„ªå‹¢èˆ‡å›°å¢ƒã€‚</p>
<p>Coding variant å¾ˆå¥½ç†è§£ï¼Œå°±æ˜¯åœ¨æŸå€‹ gene coding region ç”¢ç”Ÿçš„åºåˆ—æ”¹è®Šï¼Œä¸€èˆ¬æœƒå…ˆçœ‹æ‰€è¬‚çš„ nonsynonymousï¼Œå³é€™å€‹ variant é€ æˆ amino acid æ”¹è®Šï¼Œå½±éŸ¿åˆ°è›‹ç™½è³ªçš„çµæ§‹ï¼Œé€²è€Œå½±éŸ¿åˆ°å…¶åŠŸèƒ½ã€‚synonymous variant é›–ç„¶ä¸æœƒæ”¹è®Š amino acidï¼Œä½†åœ¨æ¨¡å¼ç‰©ç¨®ä¸­ï¼Œå¯èƒ½æœƒè¨è«–ä¸åŒ amino acid å°æ–¼ä¸åŒ tRNA çš„åå¥½ï¼Œä¹Ÿè¨±æœƒå½±éŸ¿åˆ° gene expressionã€‚å¦ä¸€æ–¹é¢ï¼Œå®ƒä¹Ÿå¯èƒ½æœƒå½±éŸ¿ transcription factor (TF) bindingï¼ŒæŸäº› TF åœ¨ biding æœ‰åå¥½çš„ DNA sequenceï¼ˆmotifï¼‰ï¼Œå³ä½¿è›‹ç™½è³ªåºåˆ—ä¸è®Šï¼ŒTF binding è®ŠåŒ–ä¹Ÿæœƒå½±éŸ¿åˆ°å…¶ä»–åŸºå› çš„èª¿æ§ã€‚</p>
<p>ä¸éä¸€èˆ¬è€Œè¨€ï¼Œcoding variant ä¸»è¦éƒ½æ˜¯è€ƒæ…® nonsynonymous changeï¼Œé€™é€ æˆçš„è®ŠåŒ–ååˆ†å…·å¤§ï¼Œç„¡æ³•è§£é‡‹åƒ complex traitsã€gene expression é«˜ä½é€™ç¨®ç´°å¾®çš„è®ŠåŒ–ã€‚</p>
<h3 id="noncoding-elements">Noncoding elements</h3>
<p>Noncoding vairant ç›¸å°è€Œè¨€è¤‡é›œçš„å¤šã€‚åœ¨è¨è«–å®ƒä¹‹å‰ï¼Œä¸å¦‚ä¾†èªªèªªçœ‹æˆ‘å€‘çŸ¥é“å“ªäº› noncoding elementsï¼š</p>
<ul>
<li>Introns</li>
<li>Promoters</li>
<li>Regulatory elements (REs)<ul>
<li><em>cis</em>-regulatory elements (CREs): promoters, enhancers</li>
<li><em>trans</em>-regulatory elements</li>
</ul>
</li>
<li>miRNAs</li>
<li>Retrovirus, satellites, centromeres, telemeres</li>
<li>Structual elements <ul>
<li>Matrix Attachment Regions (MARs)</li>
<li>Lamina Associated Domains (LADs)</li>
<li>CTCF/Cohesin</li>
<li>Topologically associating domains (TADs)</li>
<li>3D genome<sup id="fnref:3D genome"><a class="footnote-ref" href="#fn:3D genome">1</a></sup></li>
</ul>
</li>
<li>Methylation</li>
</ul>
<p>å’¦ï¼Œå¿˜è¨˜æåˆ° histone modification å—ï¼Ÿé—œæ–¼é€™äº› epigenetics markersï¼ŒBarak å°æ–¼ä»–å€‘æœ‰æ·±åˆ»çš„æ‡·ç–‘ï¼Œä»–èªç‚ºé€™äº›åªæ˜¯ markers è€Œéæœ€çµ‚ regulatory elementï¼š</p>
<blockquote>
<p>&ldquo;Something you can measure does not mean it is interesting.&rdquo;</p>
</blockquote>
<p>ç„¶å¾Œå»ºè­°æˆ‘å€‘å»è®€ä¸€ç¯‡æ‰¹è©• ENCODE çš„è«–æ–‡<sup id="fnref:ENCODE paper"><a class="footnote-ref" href="#fn:ENCODE paper">2</a></sup>ï¼Œè¢«ä»–è©•ä¹‹ç‚ºè¿‘åå¹´æœ€è¾›è¾£ï¼Œæ¨™é¡Œä¹Ÿéå¸¸æœ‰è¶£ã€‚</p>
<h3 id="noncoding-variants">Noncoding variants</h3>
<p>å¾ non-coding elements æˆ‘å€‘èƒ½çŸ¥é“æ§åˆ¶ gene expression å¯ä»¥å¾å¾ˆå¤šé¢å‘åˆ‡å…¥ï¼Œæ–¼æ˜¯è¨è«– non-coding variant æ™‚å°±æœƒæœ‰å¾ˆå¤šä¸åŒçš„æ©Ÿåˆ¶å½±éŸ¿ gene expressionã€‚åº•ä¸‹é‡å°æ‰€è¬‚çš„ enhancers (RE) å’Œ promoters ä¾†ç•«å€‹ç°¡å–®çš„ç¤ºæ„åœ–ï¼š</p>
<div class="highlight"><pre><span></span><code class="language-text">     TF1  TF2  TF3         TF4  [RNA PolII]
----[  enhancer  ]-------[promoter]--[gene body]-------------------------
  &lt;-- Topologically Associating Domain, TAD -----&gt;   &lt;-- Another TAD --&gt;
</code></pre></div>

<p>RNA Polymerase II (RNA PolII) è² è²¬ gene transcriptionï¼Œè€Œ promoter æ˜¯ä¸€æ®µåœ¨ gene body å‰ä¸ç‰¹å®šçš„åºåˆ—å¯ä»¥å¸å¼• RNA PolII ä¾†æé«˜ gene transcription rateï¼Œå¾ˆæœ‰å¯èƒ½å°±æœƒæé«˜ gene expressionã€‚TF å¯èƒ½æœƒè¾¨åˆ¥ promoter ä¸Šç‰¹å®šçš„åºåˆ—ï¼Œå®ƒå° PolII æœ‰æ›´å¼·çš„å¸å¼•åŠ›ã€‚é™¤äº† promoter ä¹‹å¤–ï¼Œenhancer ç›¸è¼ƒæ–¼ gene body çš„è·é›¢å°±æ›´ä¸ç¢ºå®šï¼Œå¯èƒ½æ˜¯ 10kb æˆ– 100kb ä¹‹å¤–ï¼Œä½†å®ƒåœ¨ç«‹é«”çš„è·é›¢å¯èƒ½éå¸¸è¿‘ï¼Œæœ¬èº«ä¹Ÿå¯ä»¥ recruit TF ç„¶å¾Œå¢åŠ  Pol affinityã€‚é€™ä¸€åˆ‡å¯ä»¥ç”¨æŠ‘åˆ¶ã€ç«¶çˆ­çš„è§’åº¦ä¾†æƒ³ç”¢ç”Ÿè² å‘çš„èª¿æ§ã€‚</p>
<p>Enhancer çš„å½±éŸ¿åŠ›æ²’æœ‰æ–¹å‘æ€§ï¼Œå³ä¸Šä¸‹æ¸¸çš„ gene éƒ½æœƒå—åŒå€‹ enhancer èª¿æ§ã€‚æ–¼æ˜¯æœ‰æ‰€è¬‚ TAD çš„æ¦‚å¿µï¼Œå®ƒæœƒè®“ chromosome å½¢æˆä¸€å€‹ loop ä¾·é™é€™æ¨£ç«‹é«”ç©ºé–“ä¸Šä¸‹æ¸¸çš„äº’å‹•ï¼Œä½¿å¾—åªåœ¨åŒå€‹ TAD çš„ REs å’Œ gene èƒ½äº’ç›¸ä½œç”¨ã€‚
é€™æ¨£çš„è§€å¿µå¯ä»¥é€²ä¸€æ­¥æ¨å»£åˆ° 3D genome ä¸Šï¼Œè€ƒæ…®ä¸åŒ chromosome é–“çš„äº’å‹•ã€‚TAD çš„é‚Šç•Œç”±æŸäº› motifï¼ˆä¾‹å¦‚ CTCFï¼‰æ±ºå®šï¼Œä½†ç©¶ç«Ÿ TAD æ˜¯å¦‚æœå»ºç«‹èˆ‡èª¿æ§ï¼Œæ©Ÿåˆ¶å°šæœªæ˜æœ—ã€‚</p>
<p>åœ¨ non-coding è¤‡é›œçš„äº¤äº’ä½œç”¨çš„å¦ä¸€é¢ï¼Œä»£è¡¨äº†æ¯å€‹äº¤äº’ä½œç”¨å¾ˆå¯èƒ½åƒ…æ”¹è®Šäº†åŸºå› è¡¨é”çš„ç¨‹åº¦ï¼Œè€Œä¸æ˜¯å¤§å¹…åº¦çš„é–‹é—œã€‚ä½†é€™ä¹Ÿä»£è¡¨ä»–å€‘å°ç”Ÿç‰©é«”ä¸ä¸€å®šæœ‰å¾ˆå¼·çš„å½±éŸ¿ï¼Œæ‰€ä»¥æœ‰è®ŠåŒ–ä¸¦ä¸€å®šä»£è¡¨å®ƒæœ‰åŠŸèƒ½ã€‚ä¸éï¼Œä¸åŒçš„ cell type å€’å¯ä»¥ç”¨é€é TF æœ‰ç„¡ä¾†èª¿æ§ä¸€ç³»åˆ—çš„ geneï¼Œè€Œä¸æ˜¯ä¸€å‘³å¢åŠ  gene æ•¸é‡ã€‚å› æ­¤ï¼Œåœ¨å¾ˆå¤šæƒ…æ³ä¸‹ï¼Œäº†è§£ non-coding variants é€ æˆçš„å½±éŸ¿æ˜¯å¾ˆæœ‰è¶£çš„ã€‚</p>
<h3 id="endophenotypes">Endophenotypes</h3>
<p>æˆ‘å€‘è¦å¦‚ä½•çœ‹ non-coding variants å‘¢ï¼Ÿé¦–å…ˆè¦äº†è§£å¾ genotype åˆ° phenotype å…¶å¯¦ä¸­é–“åŒ…å«äº†å¾ˆå¤šå±¤ç´šï¼š</p>
<div class="highlight"><pre><span></span><code class="language-text">DNA (genotype) â†’  RNA â†’  Proteins â†’  Metabolites â†’  Phenotype
</code></pre></div>

<p>ä¸­é–“çš„æ¯å€‹æ­¥é©Ÿéƒ½å¯èƒ½å½±éŸ¿ï¼Œæˆ–ä¸æœƒå‚³éå½±éŸ¿è‡³ä¸‹å€‹éšæ®µã€‚ä½†æˆ‘å€‘åœ¨ DNA å’Œ RNA level æœ‰éå¸¸å¥½çš„å·¥å…·â”€å®šåºâ”€å¯ä»¥åŒæ™‚çœ‹ genome wide éå¸¸å¤šåŸºå› æˆ–å€åŸŸã€‚æ–¼æ˜¯åœ¨å¤§å¤šæ•¸çš„æƒ…æ³æˆ‘å€‘éƒ½åªæœ‰çœ‹åˆ° endophenotypesï¼Œè¦å‹™å¿…åƒ…è¨˜åœ¨å¿ƒé€™å’ŒçœŸæ­£çš„ phenotype æ˜¯æœ‰æ‰€å·®ç•°çš„ã€‚</p>
<h2 id="eqtl">eQTL</h2>
<p>eQTL å³æ˜¯ä¸€ç¨® endophenotypeã€‚QTL (quantitative trait loci) æ„å³æŸå€‹ chromosome region å¯ä»¥é—œé€£è‡³ä¸€äº›é‡åŒ–æ•¸å€¼çš„è®Šå‹•ï¼ˆå³ locations that map to some quantitative measuresï¼‰ï¼Œè€Œ eQTL å³ç‚º expresion QTLï¼Œé—œå¿ƒæŸæ®µå€åŸŸçš„ variants å½±éŸ¿ gene expressionã€‚</p>
<p>éå¾€å¸¸è¦‹çš„ eQTL study æœ‰ï¼š</p>
<ul>
<li>Linkage study</li>
<li>Family tree</li>
<li>GWAS on two groups</li>
</ul>
<p>é€™éƒ½æ˜¯ä½¿ç”¨æ•¸å€‹ä¸åŒäººä¸åŒ sample ä¾†çœ‹ eQTLã€‚ä½†é€™äº›å°æ–¼ non-coding variant ä¾†èªªè®Šå› å¤ªå¤šï¼Œç‚ºä½•ä¸å¾å€‹äººã€å–®ä¸€ sample è‘—æ‰‹ï¼Œå³ allele imballanceï¼Ÿç„¶è€Œå–®ä¸€ sample å°±æœƒç‰½æ‰¯åˆ° eQTL æœ¬èº«çš„å•é¡Œï¼Œå³å®ƒå¾ˆé›£é€²ä¸€æ­¥å¾æŸå€‹å€åŸŸç¸®å°åˆ°æ˜¯å“ªå€‹ variant æˆ–å“ªå¹¾å€‹ variants ç‚ºæ±ºå®šæ€§å› å­ (causal vairants)ã€‚</p>
<h2 id="mpra">MPRA</h2>
<p>æ–¼æ˜¯æˆ‘å€‘å¯ä»¥æƒ³è¾¦æ³•è¨­è¨ˆå¯¦é©—ä¾†é€²ä¸€æ­¥è§£é‡‹ eQTLã€‚å¯¦é©—å¯ä»¥å¾å…©å€‹æ–¹å‘ä¾†è¨­è¨ˆï¼šnecessary å’Œ sufficientã€‚Necessity å¯ä»¥é€é CRISPR è¨­è¨ˆä¸€ç³»åˆ—çš„ tiled gRNAs æŠŠæŸå€‹ eQTL é€æ­¥åˆªæ‰ã€‚å¹³è¡ŒåŒ–é€™å€‹å¯¦é©—ï¼Œå¯ä»¥é€é growth selection å’Œ single cell sequencing è®€å‡ºæ˜¯å“ªäº› gRNAs æœ€æœ‰å½±éŸ¿åŠ›ã€‚</p>
<p>åœ¨ sufficiency æ–¹é¢ï¼Œæˆ‘å€‘å¯ä»¥è¨­è¨ˆ reporter assay ä¾†å›ç­”é€™å•é¡Œï¼š</p>
<div class="highlight"><pre><span></span><code class="language-text">-----------------------------[weak promoter]--[GFP]--
---[cis RE, CRE]-------------[weak promoter]--[GFP]--
</code></pre></div>

<p>reporter assay å¯ä»¥ç”¨å€‹ plasmid æ”¾åˆ° target cellï¼Œä½†è¦æ€éº¼å¹³è¡ŒåŒ–ï¼ŒåŒæ™‚çœ‹å¾ˆå¤š genes å‘¢ï¼Ÿé€™æ™‚å€™å°±æ˜¯ MPRA (Massively Parallel Reporter Assay) è¡¨ç¾çš„æ™‚å€™äº†ã€‚æˆ‘å€‘å¯ä»¥ç”¨ DNA synthesis æŠŠè©² <em>cis</em>-regulatory element (CRE) å’Œ barcode åšå‡ºä¾†ï¼Œå¯ä»¥å»ºç«‹ä¸€å€‹ CRE libraryï¼Œç”¨ RNA-seq å°±å¯ä»¥åŒæ™‚çœ‹åˆ°ä¸åŒ CRE æ‰€é€ æˆçš„ gene expression changeã€‚ç•¶ç„¶ç´°ç¯€æœ‰åƒ normalization DNA amount å’Œ barcode efficiencyï¼Œä½†æˆ‘å€‘å¯ä»¥ç”¨ MPRA ä¾†åˆ†æ CREã€‚</p>
<p>é€™è£¡æåˆ°çš„ CRE-seq æœ‰ä»€éº¼ç¼ºé™·å‘¢ï¼Ÿå®ƒæ˜¯ Plasmid basedï¼Œæ²’æœ‰ histone modificationï¼Œæœ‰ copy number å•é¡Œï¼›å†ä¾†ä»–çš„ genome context ä¹Ÿåªæœ‰å€åŸŸæ€§ï¼ˆåƒ TAD å°±æ²’æœ‰è€ƒæ…®ï¼‰ã€‚æ–¼æ˜¯æ¥ä¸‹ä¾†å¦‚ä½•æ”¹å–„ä»–ï¼Œå°±æ˜¯ç›®å‰ Barak Lab ç ”ç©¶æœ€æ–°å‹•æ…‹ã€‚</p>
<h2 id="conclusion">Conclusion</h2>
<p>æˆ‘è¦ºå¾—å¾é€™å€‹è§’åº¦ï¼ŒæŠŠå¾ˆå¤šè§€å¿µç”¨ç³»çµ±çš„è§’åº¦æ•´åˆï¼Œä¸¦ä¸”æå‡ºæ–°çš„å¯¦é©—èˆ‡æ¨¡å‹ï¼Œéå¸¸æœ‰è¶£ã€‚åƒè¦æ€éº¼ model enhancer å’Œ TFs çš„äº¤äº’ä½œç”¨ï¼Œéƒ½æ˜¯å¾ˆæœ‰è¶£çš„é¡Œç›®ã€‚ä»–çš„èª²éå¸¸æœ‰å•Ÿç™¼æ€§ï¼Œå¾ˆæœ‰æ„æ€ã€‚</p>
<div class="footnote">
<hr>
<ol>
<li id="fn:3D genome">
<p>é—œæ–¼ 3D genome å°±è¦æä¸€ä¸‹é€™ç¯‡è«–æ–‡ï¼š<br>
Adrian and Suhas <em>el al.</em>, <a href="http://www.pnas.org/content/112/47/E6456.abstract"><em>Chromatin extrusion explains key features of loop and domain formation in wild-type and engineered genomes</em></a>, PNAS, 2015.<br>
è£¡é¢ç”¨ç¢å½¢ (fractal globule) å»è§£é‡‹ CTCF å½¢æˆ TADs é€ æˆæ€éº¼æ¨£çš„æŸ“è‰²é«”æ‘ºç–Šï¼Œä¸¦å¦‚ä½•é€éé€™æ¨£çš„æ‘ºç–Šç”¢ç”Ÿ long distance interactionï¼Œå› ç‚ºå¯èƒ½åœ¨ç«‹é«”ç©ºé–“ä»–å€‘æ˜¯æ¥è¿‘çš„ã€‚æ¨¡å‹ç”¨ä¾†è§£é‡‹ Hi-C æ•¸æ“šã€‚é€™ç¯‡è«–æ–‡ä½¿ç”¨æ•¸å­¸ä¹‹æŠ½è±¡å’Œè¤‡é›œï¼Œç”šè‡³è«‹ä¸˜æˆæ¡ä¾†ç•¶ reviewerã€‚&#160;<a class="footnote-backref" href="#fnref:3D genome" title="Jump back to footnote 1 in the text">&#8617;</a></p>
</li>
<li id="fn:ENCODE paper">
<p>é‡å° ENCODE æ‰€è¬‚ 80% genome are functional éå¸¸æœ‰åçš„æˆ°æ–‡ï¼š<br>Dan <em>et al.</em>, <a href="https://academic.oup.com/gbe/article-lookup/doi/10.1093/gbe/evt028"><em>On the Immortality of Television Sets: â€œFunctionâ€ in the Human Genome According to the Evolution-Free Gospel of ENCODE</em></a>, Genome Biol Evol, 2013.&#160;<a class="footnote-backref" href="#fnref:ENCODE paper" title="Jump back to footnote 2 in the text">&#8617;</a></p>
</li>
</ol>
</div></section>
    <footer class="post-footer">
	<section
		class="utterances"
		data-repo="ccwang002/ccwang002.github.io"
		data-issue-term="pathname"
		data-label="Comment">
	</section>
	</footer>
</article>
	</main>
	<footer class="site-footer">
		<a class="subscribe"
			href="https://blog.liang2.tw/feeds/all.atom.xml"
			data-tooltip="Atom feed"
		>
			<svg viewBox="0 0 24 24" width="24" height="24" alt="Atom feed">
				<circle cx="6.18" cy="17.82" r="2.18"/>
				<path d="m4 10.1v2.83c3.9 0 7.07 3.17 7.07 7.07h2.83c0-5.47-4.43-9.9-9.9-9.9zm0-5.66v2.83c7.03 0 12.73 5.7 12.73 12.73h2.83c0-8.59-6.97-15.56-15.56-15.56z"/>
			</svg>
		</a>
		<div class="inner">
			<section>
				Browse articles by
				<a href="https://blog.liang2.tw/tags.html">Tags</a> |
				<a href="https://blog.liang2.tw/categories.html">Categories</a>
			</section>
			<section class="contact">
				Contact me by
				<a href="https://twitter.com/lbwang2">Twitter</a> |
				<a href="https://www.facebook.com/lbwang.2">Facebook</a> |
				<a href="https://github.com/ccwang002">GitHub</a> |
				<a href="mailto:me+blog@liang2.tw">Email</a> |
				<a href="https://www.linkedin.com/in/liangbowang/">LinkedIn</a>
			</section>
			<section class="copyright">
				This work is licensed under <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.
			</section>
			<section class="poweredby">
				Built using <a href="https://getpelican.com/">Pelican</a>.
				Powered by the <a href="https://kura.github.io/hauntr/">Hauntr</a> theme.
			</section>
		</div>
	</footer>
	<script src="https://blog.liang2.tw/theme/scripts/theme-toggle.js"></script>
</body>
</html>