<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="utf-8" />
	<meta http-equiv="X-UA-Compatible" content="IE=edge, chrome=1">
	<meta http-equiv="content-type" content="text/html; charset=UTF-8;charset=utf-8">
	<meta name="viewport" content="width=device-width,initial-scale=1, maximum-scale=1">
	<meta name="author" content="Liang-Bo Wang" />
	<meta name="description" content="" />
	<link rel="shortcut icon" href="https://blog.liang2.tw/favicon.ico" />
	<title>Liang-Bo Wang's Blog</title>

	<!-- og -->
	<meta property="og:locale" content="en_US"/>
	<meta property="og:site_name" content="Liang-Bo Wang's Blog"/>
	<meta property="og:type" content="blog"/>
	<meta property="og:title" content="Liang-Bo Wang's Blog"/>
	<meta property="og:description" content=""/>
	<meta property="og:url" content="https://blog.liang2.tw"/>

	<!-- Feeds -->
	<link href="https://blog.liang2.tw/feeds/all.atom.xml"
		type="application/atom+xml" rel="alternate" title="Liang-Bo Wang's Blog" />

	<link rel="stylesheet" type="text/css" href="https://blog.liang2.tw/theme/css/style.css" />
	<link rel="stylesheet" type="text/css" href="https://blog.liang2.tw/theme/css/pygments.css" />
	<link rel="stylesheet" type="text/css" href="https://blog.liang2.tw/theme/katex-0.11.1/katex.min.css">
</head>

<body class="post-template">
	<header class="site-head">
		<h1 class="blog-title"><a href="https://blog.liang2.tw">Liang-Bo Wang's Blog</a></h1>
		<h2 class="blog-description">
			<a href="https://blog.liang2.tw/about-me/">About</a> |
			<a href="https://blog.liang2.tw/talks/">Talks</a> |
			<a href="https://blog.liang2.tw/archives.html">Archives</a>
		</h2>
	</header>
	<main id="content" class="content" role="main">
    <article class="post">
        <header class="post-header">
            <span class="post-meta">
                <time datetime="Jul 24, 2020">Jul 24, 2020</time>
                in <a href="https://blog.liang2.tw/category/bioinfo/">Bioinfo</a>
            </span>
            <h2 class="post-title"><a href="https://blog.liang2.tw/posts/2020/07/id-crossref-exact-protein-uniparc/" rel="bookmark" title="Permalink to ID cross reference with exact protein sequence identity using UniParc">ID cross reference with exact protein sequence identity using UniParc</a></h2>
            <span class="post-meta">
                        <a href="https://blog.liang2.tw/tag/en/">en</a>
                        <a href="https://blog.liang2.tw/tag/id-crossref/">id-crossref</a>
                        <a href="https://blog.liang2.tw/tag/cptac/">cptac</a>
            </span>
        </header>
        <section class="post-excerpt">
            <p><em>TL;DR: Many existing ID mappings between different biological ID systems (RefSeq/Ensembl/UniProt) don&rsquo;t consider if the IDs have the same exact protein sequence. When the exact sequence is needed, UniParc can be used to cross-reference the IDs. I will demonstrate how to use UniParc to map RefSeq …</em></p>
        </section>
    </article>
    <article class="post">
        <header class="post-header">
            <span class="post-meta">
                <time datetime="Jun 08, 2020">Jun 08, 2020</time>
                in <a href="https://blog.liang2.tw/category/bioinfo/">Bioinfo</a>
            </span>
            <h2 class="post-title"><a href="https://blog.liang2.tw/posts/2020/06/identify-ensembl-release-ver/" rel="bookmark" title="Permalink to Identify the Ensembl release from versioned IDs">Identify the Ensembl release from versioned IDs</a></h2>
            <span class="post-meta">
                        <a href="https://blog.liang2.tw/tag/en/">en</a>
            </span>
        </header>
        <section class="post-excerpt">
            <p>I often received data that was annotated by an unknown Ensembl release/version.</p>
<p>It could be the Ensembl IDs in a gene expression matrix, a VEP annotated MAF file, or even a customized GTF. The documentation of those files wasn&rsquo;t always clear about the annotation in use. However, it …</p>
        </section>
    </article>
    <article class="post">
        <header class="post-header">
            <span class="post-meta">
                <time datetime="Apr 20, 2019">Apr 20, 2019</time>
                in <a href="https://blog.liang2.tw/category/bioinfo/">Bioinfo</a>
            </span>
            <h2 class="post-title"><a href="https://blog.liang2.tw/posts/2019/04/generate-venn-svg/" rel="bookmark" title="Permalink to Generate Venn diagrams easily">Generate Venn diagrams easily</a></h2>
            <span class="post-meta">
                        <a href="https://blog.liang2.tw/tag/en/">en</a>
            </span>
        </header>
        <section class="post-excerpt">
            <p>I find myself generating Venn diagrams quite often.  While there are many available Venn diagram plotting libraries available, they don&rsquo;t always fit my need.  My inputs of the diagram are the set sizes rather than lists of observations.  And after drawing the Venn diagram, I often edit them to …</p>
        </section>
    </article>
    <article class="post">
        <header class="post-header">
            <span class="post-meta">
                <time datetime="Feb 26, 2019">Feb 26, 2019</time>
                in <a href="https://blog.liang2.tw/category/bioinfo/">Bioinfo</a>
            </span>
            <h2 class="post-title"><a href="https://blog.liang2.tw/posts/2019/02/gdc-seqinfo/" rel="bookmark" title="Permalink to Store GDC genome as a Seqinfo object">Store GDC genome as a Seqinfo object</a></h2>
            <span class="post-meta">
                        <a href="https://blog.liang2.tw/tag/en/">en</a>
                        <a href="https://blog.liang2.tw/tag/r/">r</a>
                        <a href="https://blog.liang2.tw/tag/bioconductor/">bioconductor</a>
            </span>
        </header>
        <section class="post-excerpt">
            <p>Genomic Data Commons (GDC) hosted by NCI is the place to harmonize past and future genomic data, such as TCGA, TARGET, and CPTAC projects. GDC has its own genome reference, <a href="https://gdc.cancer.gov/about-data/data-harmonization-and-generation/gdc-reference-files"><code>GRCh38.d1.vd1</code></a>, which has 2,779 &ldquo;chromosomes&rdquo; including decoys and virus sequences. That said, the canonical chromosomes of GRCh38 …</p>
        </section>
    </article>
    <article class="post">
        <header class="post-header">
            <span class="post-meta">
                <time datetime="Jan 08, 2019">Jan 08, 2019</time>
                in <a href="https://blog.liang2.tw/category/bioinfo/">Bioinfo</a>
            </span>
            <h2 class="post-title"><a href="https://blog.liang2.tw/posts/2019/01/build-ensdb-from-local-mysql/" rel="bookmark" title="Permalink to Build EnsDb from a local Ensembl MySQL database">Build EnsDb from a local Ensembl MySQL database</a></h2>
            <span class="post-meta">
                        <a href="https://blog.liang2.tw/tag/en/">en</a>
                        <a href="https://blog.liang2.tw/tag/r/">r</a>
                        <a href="https://blog.liang2.tw/tag/bioconductor/">bioconductor</a>
                        <a href="https://blog.liang2.tw/tag/sqlite/">sqlite</a>
                        <a href="https://blog.liang2.tw/tag/ensembldb/">ensembldb</a>
            </span>
        </header>
        <section class="post-excerpt">
            <p>In some occasions, I need to access the older version of Ensembl human transcripts. For example, the mutation calls generated by the <a href="https://gdc.cancer.gov/">NCI&rsquo;s Genomic Data Common</a> pipeline are annotated by Ensembl v84. To programmatically query the Ensembl annotations, I use the EnsDb SQLite database created by <a href="https://bioconductor.org/packages/release/bioc/html/ensembldb.html">ensembldb</a>, which is …</p>
        </section>
    </article>
    <article class="post">
        <header class="post-header">
            <span class="post-meta">
                <time datetime="Dec 13, 2018">Dec 13, 2018</time>
                in <a href="https://blog.liang2.tw/category/coding/">Coding</a>
            </span>
            <h2 class="post-title"><a href="https://blog.liang2.tw/posts/2018/12/firefox-borderless-fullscreen-macos/" rel="bookmark" title="Permalink to Make Firefox fullscreen borderless on macOS">Make Firefox fullscreen borderless on macOS</a></h2>
            <span class="post-meta">
                        <a href="https://blog.liang2.tw/tag/en/">en</a>
            </span>
        </header>
        <section class="post-excerpt">
            <p>EDIT 2021-06-01: In Firefox 89+, there&rsquo;s a default option &ldquo;Hide Toolbar&rdquo; in the fullscreen mode that automatically hides the toolbar. So the customization is no longer needed.</p>
<p>Firefox fullscreen on macOS by default contains the address bar and the tab bar. I usually don&rsquo;t really need the full …</p>
        </section>
    </article>
    <article class="post">
        <header class="post-header">
            <span class="post-meta">
                <time datetime="Oct 20, 2018">Oct 20, 2018</time>
                in <a href="https://blog.liang2.tw/category/coding/">Coding</a>
            </span>
            <h2 class="post-title"><a href="https://blog.liang2.tw/posts/2018/10/gpg-key-transition-2018/" rel="bookmark" title="Permalink to GPG Key Transition">GPG Key Transition</a></h2>
            <span class="post-meta">
                        <a href="https://blog.liang2.tw/tag/en/">en</a>
                        <a href="https://blog.liang2.tw/tag/gpg/">gpg</a>
            </span>
        </header>
        <section class="post-excerpt">
            <p>I am transiting my GPG key again.  However, for this time, I expect to use the new GPG master key longer and will start building this identity unless there is a concern about the key strength or I accidentally lose the key. </p>
<p>Back in <a href="https://blog.liang2.tw/posts/2016/12/gpg-key-transition-2016/">my GPG key transition in 2016 …</a></p>
        </section>
    </article>
    <article class="post">
        <header class="post-header">
            <span class="post-meta">
                <time datetime="Jun 22, 2018">Jun 22, 2018</time>
                in <a href="https://blog.liang2.tw/category/bioinfo/">Bioinfo</a>
            </span>
            <h2 class="post-title"><a href="https://blog.liang2.tw/posts/2018/06/gene-annotation-using-gffutils/" rel="bookmark" title="Permalink to Access gene annotation using gffutils">Access gene annotation using gffutils</a></h2>
            <span class="post-meta">
                        <a href="https://blog.liang2.tw/tag/en/">en</a>
                        <a href="https://blog.liang2.tw/tag/python/">python</a>
                        <a href="https://blog.liang2.tw/tag/sqlite/">sqlite</a>
            </span>
        </header>
        <section class="post-excerpt">
            <p>Recently, I had to access gene annotations in multiple versions from multiple sources such as Ensembl, GENCODE, and UCSC. I used to rely on the R/Bioconductor ecosystem to query the coordinates of a gene annotation. There are existing Bioconductor packages ready for Ensembl and UCSC annotations (more info in …</p>
        </section>
    </article>
    <article class="post">
        <header class="post-header">
            <span class="post-meta">
                <time datetime="Jan 28, 2018">Jan 28, 2018</time>
                in <a href="https://blog.liang2.tw/category/bioinfo/">Bioinfo</a>
            </span>
            <h2 class="post-title"><a href="https://blog.liang2.tw/posts/2018/01/read-uniprotkb-xml/" rel="bookmark" title="Permalink to Read UniProtKB in XML format">Read UniProtKB in XML format</a></h2>
            <span class="post-meta">
                        <a href="https://blog.liang2.tw/tag/en/">en</a>
                        <a href="https://blog.liang2.tw/tag/python/">python</a>
                        <a href="https://blog.liang2.tw/tag/uniprot/">uniprot</a>
            </span>
        </header>
        <section class="post-excerpt">
            <p>UniProt Knowledge Base (<a href="http://www.uniprot.org/help/uniprotkb">UniProtKB</a>) provides <a href="https://www.uniprot.org/help/programmatic_access">various methods</a> to access their data. I settled on their XML format since no additional parsing code is required and the format is well defined, which comes with a schema. Plus, it turns out that databases such as <a href="http://pdbml.pdb.org/">PDB</a> also provide their data export in …</p>
        </section>
    </article>
    <article class="post">
        <header class="post-header">
            <span class="post-meta">
                <time datetime="Jan 20, 2018">Jan 20, 2018</time>
                in <a href="https://blog.liang2.tw/category/bioinfo/">Bioinfo</a>
            </span>
            <h2 class="post-title"><a href="https://blog.liang2.tw/posts/2018/01/ad-hoc-bioinfo-analysis-in-database/" rel="bookmark" title="Permalink to Ad hoc bioinformatic analysis in database">Ad hoc bioinformatic analysis in database</a></h2>
            <span class="post-meta">
                        <a href="https://blog.liang2.tw/tag/en/">en</a>
                        <a href="https://blog.liang2.tw/tag/python/">python</a>
                        <a href="https://blog.liang2.tw/tag/pandas/">pandas</a>
                        <a href="https://blog.liang2.tw/tag/sqlite/">sqlite</a>
                        <a href="https://blog.liang2.tw/tag/postgresql/">postgresql</a>
            </span>
        </header>
        <section class="post-excerpt">
            <p>Recently I&rsquo;ve found that bioinformatic analysis in a database is not hard at all and the database set up wasn&rsquo;t as daunting as it sounds, especially when the data are tabular. I used to start my analysis with loading everything into R or Python, and then figuring out …</p>
        </section>
    </article>
<nav class="pagination" role="navigation">
		<span class="page-number">Page 1 of 5</span>
	<a class="older-posts" href="https://blog.liang2.tw/index2.html">Next &rarr;</a>
</nav>
	</main>
	<footer class="site-footer">
		<a class="subscribe icon-feed" href="https://blog.liang2.tw/feeds/all.atom.xml"><span class="tooltip">Atom</span></a>
		<div class="inner">
			<section>
				Browse articles by
				<a href="https://blog.liang2.tw/tags.html">Tags</a> |
				<a href="https://blog.liang2.tw/categories.html">Categories</a>
			</section>
			<section class="contact">
				Contact me by
				<a href="https://twitter.com/lbwang2">Twitter</a> |
				<a href="https://www.facebook.com/lbwang.2">Facebook</a> |
				<a href="https://github.com/ccwang002">GitHub</a> |
				<a href="mailto:me+blog@liang2.tw">Email</a> |
				<a href="https://www.linkedin.com/in/liangbowang/">LinkedIn</a>
			</section>
			<section class="copyright">
				This work is licensed under <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.
			</section>
			<section class="poweredby">
				Built using <a href="https://getpelican.com/">Pelican</a>.
				Powered by the <a href="https://kura.github.io/hauntr/">Hauntr</a> theme.
			</section>
		</div>
	</footer>
</body>
</html>