<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="utf-8" />
	<meta http-equiv="X-UA-Compatible" content="IE=edge, chrome=1">
	<meta http-equiv="content-type" content="text/html; charset=UTF-8;charset=utf-8">
	<meta name="viewport" content="width=device-width,initial-scale=1, maximum-scale=1">
<meta name="author" content="Liang-Bo Wang" />
<meta name="description" content="Genomic Data Commons (GDC) hosted by NCI is the place to harmonize past and future genomic data, such as TCGA, TARGET, and CPTAC projects. GDC has its own genome reference, GRCh38.d1.vd1, which has 2,779 “chromosomes” including decoys and virus sequences. That said, the canonical chromosomes of GRCh38 …" />
<meta name="keywords" content="en, r, bioconductor">
	<link rel="icon" href="data:image/svg+xml,<svg xmlns=%22http://www.w3.org/2000/svg%22 viewBox=%220 0 100 100%22><text y=%22.95em%22 font-size=%2295%22>📖</text></svg>">
	<title>Store GDC genome as a Seqinfo object</title>

	<!-- og -->
<meta property="og:locale" content="en_US"/>
<meta property="og:site_name" content="Liang-Bo Wang's Blog"/>
<meta property="og:title" content="Store GDC genome as a Seqinfo object"/>
<meta property="og:description" content="Genomic Data Commons (GDC) hosted by NCI is the place to harmonize past and future genomic data, such as TCGA, TARGET, and CPTAC projects. GDC has its own genome reference, GRCh38.d1.vd1, which has 2,779 “chromosomes” including decoys and virus sequences. That said, the canonical chromosomes of GRCh38 …"/>
<meta property="og:locale" content=""/>
<meta property="og:url" content="https://blog.liang2.tw/posts/2019/02/gdc-seqinfo/"/>
<meta property="og:type" content="article"/>
<meta property="article:published_time" content="2019-02-26 00:00:00-06:00"/>
<meta property="article:modified_time" content="2022-02-20 20:07:45.305382-06:00"/>
<meta property="article:author" content="https://blog.liang2.tw/author/liang-bo-wang.html">
<meta property="article:section" content="Bioinfo"/>
<meta property="article:tag" content="en"/>
<meta property="article:tag" content="r"/>
<meta property="article:tag" content="bioconductor"/>
<meta name="twitter:card" content="summary">
<meta name="twitter:site" content="@ccwang002">
<meta name="twitter:title" content="Store GDC genome as a Seqinfo object">
<meta name="twitter:description" content="Genomic Data Commons (GDC) hosted by NCI is the place to harmonize past and future genomic data, such as TCGA, TARGET, and CPTAC projects. GDC has its own genome reference, GRCh38.d1.vd1, which has 2,779 “chromosomes” including decoys and virus sequences. That said, the canonical chromosomes of GRCh38 …">

	<!-- Feeds -->
	<link href="https://blog.liang2.tw/feeds/all.atom.xml"
		type="application/atom+xml" rel="alternate" title="Liang-Bo Wang's Blog" />

	<link rel="stylesheet" type="text/css" href="https://blog.liang2.tw/theme/styles/styles.css" />
	<link rel="stylesheet" type="text/css" href="https://blog.liang2.tw/theme/katex-0.15.3/katex.min.css">
</head>

<body class="post-template">
	<header class="site-head">
		<h1 class="blog-title"><a href="https://blog.liang2.tw">Liang-Bo Wang's Blog</a></h1>
		<h2 class="blog-description">
			<a href="https://blog.liang2.tw/about-me/">About</a> |
			<a href="https://blog.liang2.tw/talks/">Talks</a> |
			<a href="https://blog.liang2.tw/archives.html">Archives</a> |
			<svg class="theme-toggle" width="24" height="24" viewBox="0 0 24 24" aria-label="auto" aria-live="polite">
				<path class="moon" fill="currentColor" d="M17.75,4.09L15.22,6.03L16.13,9.09L13.5,7.28L10.87,9.09L11.78,6.03L9.25,4.09L12.44,4L13.5,1L14.56,4L17.75,4.09M21.25,11L19.61,12.25L20.2,14.23L18.5,13.06L16.8,14.23L17.39,12.25L15.75,11L17.81,10.95L18.5,9L19.19,10.95L21.25,11M18.97,15.95C19.8,15.87 20.69,17.05 20.16,17.8C19.84,18.25 19.5,18.67 19.08,19.07C15.17,23 8.84,23 4.94,19.07C1.03,15.17 1.03,8.83 4.94,4.93C5.34,4.53 5.76,4.17 6.21,3.85C6.96,3.32 8.14,4.21 8.06,5.04C7.79,7.9 8.75,10.87 10.95,13.06C13.14,15.26 16.1,16.22 18.97,15.95M17.33,17.97C14.5,17.81 11.7,16.64 9.53,14.5C7.36,12.31 6.2,9.5 6.04,6.68C3.23,9.82 3.34,14.64 6.35,17.66C9.37,20.67 14.19,20.78 17.33,17.97Z" />
				<path class="sun" fill="currentColor" d="M3.55 19.09L4.96 20.5L6.76 18.71L5.34 17.29M12 6C8.69 6 6 8.69 6 12S8.69 18 12 18 18 15.31 18 12C18 8.68 15.31 6 12 6M20 13H23V11H20M17.24 18.71L19.04 20.5L20.45 19.09L18.66 17.29M20.45 5L19.04 3.6L17.24 5.39L18.66 6.81M13 1H11V4H13M6.76 5.39L4.96 3.6L3.55 5L5.34 6.81L6.76 5.39M1 13H4V11H1M13 20H11V23H13"></path>
			</svg>
		</h2>
	</header>
	<main id="content" class="content" role="main">
<article class="post">
    <span class="post-meta">
        <time datetime="2019-02-26">Feb 26, 2019</time>
        in <a href="https://blog.liang2.tw/category/bioinfo/">Bioinfo</a>
    </span>
    <h1 class="post-title">
        <a href="https://blog.liang2.tw/posts/2019/02/gdc-seqinfo/" rel="bookmark" title="Permalink to Store GDC genome as a Seqinfo object">Store GDC genome as a Seqinfo object</a>
    </h1>
    <span class="post-meta">
                <a href="https://blog.liang2.tw/tag/en/">en</a>
                <a href="https://blog.liang2.tw/tag/r/">r</a>
                <a href="https://blog.liang2.tw/tag/bioconductor/">bioconductor</a>
    </span>
    <section class="post-content"><p>Genomic Data Commons (GDC) hosted by NCI is the place to harmonize past and future genomic data, such as TCGA, TARGET, and CPTAC projects. GDC has its own genome reference, <a href="https://gdc.cancer.gov/about-data/data-harmonization-and-generation/gdc-reference-files"><code>GRCh38.d1.vd1</code></a>, which has 2,779 &ldquo;chromosomes&rdquo; including decoys and virus sequences. That said, the canonical chromosomes of GRCh38.d1.vd1 (e.g., chr1 to chr22, chrM, chrX, and chrY) are identical to that of hg38 and GRCh38. So all these three genome references can be used interchangeably.</p>
<p>Anyway, I was trying to correctly store the full GRCh38.d1.vd1 genome information in the <code>GRanges</code> and <code>GRangesList</code> R objects, which can be done by creating a <code>Seqinfo</code> object representing all its chromosomes. It was also fun to get familiar with the genomic data structures in R. </p>
<h3 id="build-gdcs-seqinfo">Build GDC&rsquo;s Seqinfo</h3>
<p>First, we need the length and the name of all chromosomes in GRCh38.d1.vd1.  I used samtools to extract the information as a <code>.dict</code> file from the genome reference FASTA file.</p>
<div class="highlight"><pre><span></span><code class="language-bash"><span class="nb">export</span> <span class="nv">GDC_REF_FA_URL</span><span class="o">=</span><span class="s1">&#39;https://api.gdc.cancer.gov/data/254f697d-310d-4d7d-a27b-27fbf767a834&#39;</span>
curl -Lo GRCh38.d1.vd1.fa.tar.gz <span class="nv">$GDC_REF_FA_URL</span>
samtools dict <span class="se">\</span>
    -a <span class="s1">&#39;GRCh38.d1.vd1&#39;</span> -s <span class="s1">&#39;Homo sapiens&#39;</span> <span class="se">\</span>
    -u <span class="nv">$GDC_REF_FA_URL</span> <span class="se">\</span>
    GRCh38.d1.vd1.fa.tar.gz &gt; GRCh38.d1.vd1.dict

head -n <span class="m">3</span> GRCh38.d1.vd1.dict
<span class="c1"># @HD   VN:1.0  SO:unsorted</span>
<span class="c1"># @SQ   SN:chr1 LN:248956422    M5:6aef897c3d6ff0c78aff06ac189178dd UR:https://api.gdc.cancer.gov/data/254f697d-310d-4d7d-a27b-27fbf767a834 AS:GRCh38.d1.vd1    SP:Homo sapiens</span>
<span class="c1"># @SQ   SN:chr2 LN:242193529    M5:f98db672eb0993dcfdabafe2a882905c UR:https://api.gdc.cancer.gov/data/254f697d-310d-4d7d-a27b-27fbf767a834 AS:GRCh38.d1.vd1    SP:Homo sapiens</span>
</code></pre></div>

<p>Seqinfo also requires the information of whether a chromosome is circular. In GDC&rsquo;s case,  mitochondria chromosome and all viruses sequences are circular. Combining all the information together, we can construct the Seqinfo object describing the genome GRCh38.d1.vd1.</p>
<div class="highlight"><pre><span></span><code class="language-splus"><span class="nf">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">GenomeInfoDb</span><span class="p">)</span>

<span class="n">gdc_simple_tbl</span> <span class="o">&lt;-</span> <span class="nf">read_tsv</span><span class="p">(</span>
    <span class="s">&#39;./GRCh38.d1.vd1.dict&#39;</span><span class="p">,</span> 
    <span class="n">skip</span> <span class="o">=</span> <span class="m">1</span><span class="p">,</span>  <span class="c1"># Skip the first line (@HQ ...)</span>
    <span class="n">col_names</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#39;SQ&#39;</span><span class="p">,</span> <span class="s">&#39;chrom&#39;</span><span class="p">,</span> <span class="s">&#39;length&#39;</span><span class="p">,</span> <span class="s">&#39;md5sum&#39;</span><span class="p">,</span> <span class="s">&#39;URI&#39;</span><span class="p">,</span> <span class="s">&#39;assembly&#39;</span><span class="p">,</span> <span class="s">&#39;species&#39;</span><span class="p">)</span>
<span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="nf">select</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">length</span><span class="p">)</span> <span class="o">%&gt;%</span>
    <span class="nf">mutate</span><span class="p">(</span><span class="n">chrom</span> <span class="o">=</span> <span class="nf">str_sub</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span> <span class="o">=</span> <span class="m">4</span><span class="p">),</span> 
           <span class="n">length</span> <span class="o">=</span> <span class="nf">as.integer</span><span class="p">(</span><span class="nf">str_sub</span><span class="p">(</span><span class="n">length</span><span class="p">,</span> <span class="n">start</span> <span class="o">=</span> <span class="m">4</span><span class="p">)),</span>
           <span class="n">circular</span> <span class="o">=</span> <span class="nf">case_when</span><span class="p">(</span>
               <span class="n">chrom</span> <span class="o">==</span> <span class="s">&#39;chrM&#39;</span> <span class="o">~</span> <span class="kc">TRUE</span><span class="p">,</span>
               <span class="n">chrom</span> <span class="o">==</span> <span class="s">&#39;chrEBV&#39;</span> <span class="o">~</span> <span class="kc">TRUE</span><span class="p">,</span>
               <span class="nf">startsWith</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="s">&#39;HPV&#39;</span><span class="p">)</span> <span class="o">~</span> <span class="kc">TRUE</span><span class="p">,</span>
               <span class="kc">TRUE</span> <span class="o">~</span> <span class="kc">FALSE</span>
           <span class="p">))</span>

<span class="n">gdc_seqinfo</span> <span class="o">&lt;-</span> <span class="nf">Seqinfo</span><span class="p">(</span>
    <span class="n">seqnames</span> <span class="o">=</span> <span class="n">gdc_simple_tbl</span><span class="o">$</span><span class="n">chrom</span><span class="p">,</span>
    <span class="n">seqlengths</span> <span class="o">=</span> <span class="n">gdc_simple_tbl</span><span class="o">$</span><span class="n">length</span><span class="p">,</span>
    <span class="n">isCircular</span> <span class="o">=</span> <span class="n">gdc_simple_tbl</span><span class="o">$</span><span class="n">circular</span><span class="p">,</span>
    <span class="n">genome</span> <span class="o">=</span> <span class="s">&#39;GRCh38.d1.vd1&#39;</span>
<span class="p">)</span>
</code></pre></div>

<p>Now we can supply it to any <code>GRanges</code> object coming out from any GDC&rsquo;s sequencing data.</p>
<div class="highlight"><pre><span></span><code class="language-rconsole"><span class="gp">&gt; </span><span class="n">gdc_seqinfo</span>
<span class="go">Seqinfo object with 2779 sequences (191 circular) from GRCh38.d1.vd1 genome:</span>
<span class="go">  seqnames   seqlengths isCircular        genome</span>
<span class="go">  chr1        248956422      FALSE GRCh38.d1.vd1</span>
<span class="go">  chr2        242193529      FALSE GRCh38.d1.vd1</span>
<span class="go">  chr3        198295559      FALSE GRCh38.d1.vd1</span>
<span class="go">  chr4        190214555      FALSE GRCh38.d1.vd1</span>
<span class="go">  chr5        181538259      FALSE GRCh38.d1.vd1</span>
<span class="go">  ...               ...        ...           ...</span>
<span class="go">  HPV-mKN2         7299       TRUE GRCh38.d1.vd1</span>
<span class="go">  HPV-mKN3         7251       TRUE GRCh38.d1.vd1</span>
<span class="go">  HPV-mL55         7177       TRUE GRCh38.d1.vd1</span>
<span class="go">  HPV-mRTRX7       7731       TRUE GRCh38.d1.vd1</span>
<span class="go">  HPV-mSD2         7300       TRUE GRCh38.d1.vd1</span>
</code></pre></div>

<p>I store the <code>gdc_seqinfo</code> as a RDS file (<a href="https://blog.liang2.tw/posts/2019/02/gdc-seqinfo/results/seqinfo_GRCh38.d1.vd1.rds">link</a> here) so I can re-use it easily.</p>
<div class="highlight"><pre><span></span><code class="language-splus"><span class="n">gdc_seqinfo</span> <span class="o">&lt;-</span> <span class="nf">readRDS</span><span class="p">(</span><span class="s">&#39;seqinfo_GRCh38.d1.vd1.rds&#39;</span><span class="p">)</span>
</code></pre></div></section>
    <footer class="post-footer">
	<section
		class="utterances"
		data-repo="ccwang002/ccwang002.github.io"
		data-issue-term="pathname"
		data-label="Comment">
	</section>
	</footer>
</article>
	</main>
	<footer class="site-footer">
		<a class="subscribe"
			href="https://blog.liang2.tw/feeds/all.atom.xml"
			data-tooltip="Atom feed"
		>
			<svg viewBox="0 0 24 24" width="24" height="24" alt="Atom feed">
				<circle cx="6.18" cy="17.82" r="2.18"/>
				<path d="m4 10.1v2.83c3.9 0 7.07 3.17 7.07 7.07h2.83c0-5.47-4.43-9.9-9.9-9.9zm0-5.66v2.83c7.03 0 12.73 5.7 12.73 12.73h2.83c0-8.59-6.97-15.56-15.56-15.56z"/>
			</svg>
		</a>
		<div class="inner">
			<section>
				Browse articles by
				<a href="https://blog.liang2.tw/tags.html">Tags</a> |
				<a href="https://blog.liang2.tw/categories.html">Categories</a>
			</section>
			<section class="contact">
				Contact me by
				<a href="https://twitter.com/lbwang2">Twitter</a> |
				<a href="https://www.facebook.com/lbwang.2">Facebook</a> |
				<a href="https://github.com/ccwang002">GitHub</a> |
				<a href="mailto:me+blog@liang2.tw">Email</a> |
				<a href="https://www.linkedin.com/in/liangbowang/">LinkedIn</a>
			</section>
			<section class="copyright">
				This work is licensed under <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.
			</section>
			<section class="powered-by">
				Built using <a href="https://getpelican.com/">Pelican</a>.
				Powered by the <a href="https://github.com/kura/hauntr">Hauntr</a> theme.
			</section>
		</div>
	</footer>
	<script src="https://blog.liang2.tw/theme/scripts/theme-toggle.js"></script>
</body>
</html>