<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="utf-8" />
	<meta http-equiv="X-UA-Compatible" content="IE=edge, chrome=1">
	<meta http-equiv="content-type" content="text/html; charset=UTF-8;charset=utf-8">
	<meta name="viewport" content="width=device-width,initial-scale=1, maximum-scale=1">
<meta name="author" content="Liang-Bo Wang" />
<meta name="description" content="In some occasions, I need to access the older version of Ensembl human transcripts. For example, the mutation calls generated by the NCIâ€™s Genomic Data Common pipeline are annotated by Ensembl v84. To programmatically query the Ensembl annotations, I use the EnsDb SQLite database created by ensembldb, which is â€¦" />
<meta name="keywords" content="en, r, bioconductor, sqlite, ensembldb">
	<link rel="icon" href="data:image/svg+xml,<svg xmlns=%22http://www.w3.org/2000/svg%22 viewBox=%220 0 100 100%22><text y=%22.95em%22 font-size=%2295%22>ðŸ“–</text></svg>">
	<title>Build EnsDb from a local Ensembl MySQL database</title>

	<!-- og -->
<meta property="og:locale" content="en_US"/>
<meta property="og:site_name" content="Liang-Bo Wang's Blog"/>
<meta property="og:title" content="Build EnsDb from a local Ensembl MySQL database"/>
<meta property="og:description" content="In some occasions, I need to access the older version of Ensembl human transcripts. For example, the mutation calls generated by the NCIâ€™s Genomic Data Common pipeline are annotated by Ensembl v84. To programmatically query the Ensembl annotations, I use the EnsDb SQLite database created by ensembldb, which is â€¦"/>
<meta property="og:locale" content=""/>
<meta property="og:url" content="https://blog.liang2.tw/posts/2019/01/build-ensdb-from-local-mysql/"/>
<meta property="og:type" content="article"/>
<meta property="article:published_time" content="2019-01-08 00:00:00-06:00"/>
<meta property="article:modified_time" content="2022-02-20 20:07:45.305268-06:00"/>
<meta property="article:author" content="https://blog.liang2.tw/author/liang-bo-wang.html">
<meta property="article:section" content="Bioinfo"/>
<meta property="article:tag" content="en"/>
<meta property="article:tag" content="r"/>
<meta property="article:tag" content="bioconductor"/>
<meta property="article:tag" content="sqlite"/>
<meta property="article:tag" content="ensembldb"/>
<meta name="twitter:card" content="summary">
<meta name="twitter:site" content="@ccwang002">
<meta name="twitter:title" content="Build EnsDb from a local Ensembl MySQL database">
<meta name="twitter:description" content="In some occasions, I need to access the older version of Ensembl human transcripts. For example, the mutation calls generated by the NCIâ€™s Genomic Data Common pipeline are annotated by Ensembl v84. To programmatically query the Ensembl annotations, I use the EnsDb SQLite database created by ensembldb, which is â€¦">

	<!-- Feeds -->
	<link href="https://blog.liang2.tw/feeds/all.atom.xml"
		type="application/atom+xml" rel="alternate" title="Liang-Bo Wang's Blog" />

	<link rel="stylesheet" type="text/css" href="https://blog.liang2.tw/theme/styles/styles.css" />
	<link rel="stylesheet" type="text/css" href="https://blog.liang2.tw/theme/katex-0.15.3/katex.min.css">
</head>

<body class="post-template">
	<header class="site-head">
		<h1 class="blog-title"><a href="https://blog.liang2.tw">Liang-Bo Wang's Blog</a></h1>
		<h2 class="blog-description">
			<a href="https://blog.liang2.tw/about-me/">About</a> |
			<a href="https://blog.liang2.tw/talks/">Talks</a> |
			<a href="https://blog.liang2.tw/archives.html">Archives</a> |
			<svg class="theme-toggle" width="24" height="24" viewBox="0 0 24 24" aria-label="auto" aria-live="polite">
				<path class="moon" fill="currentColor" d="M17.75,4.09L15.22,6.03L16.13,9.09L13.5,7.28L10.87,9.09L11.78,6.03L9.25,4.09L12.44,4L13.5,1L14.56,4L17.75,4.09M21.25,11L19.61,12.25L20.2,14.23L18.5,13.06L16.8,14.23L17.39,12.25L15.75,11L17.81,10.95L18.5,9L19.19,10.95L21.25,11M18.97,15.95C19.8,15.87 20.69,17.05 20.16,17.8C19.84,18.25 19.5,18.67 19.08,19.07C15.17,23 8.84,23 4.94,19.07C1.03,15.17 1.03,8.83 4.94,4.93C5.34,4.53 5.76,4.17 6.21,3.85C6.96,3.32 8.14,4.21 8.06,5.04C7.79,7.9 8.75,10.87 10.95,13.06C13.14,15.26 16.1,16.22 18.97,15.95M17.33,17.97C14.5,17.81 11.7,16.64 9.53,14.5C7.36,12.31 6.2,9.5 6.04,6.68C3.23,9.82 3.34,14.64 6.35,17.66C9.37,20.67 14.19,20.78 17.33,17.97Z" />
				<path class="sun" fill="currentColor" d="M3.55 19.09L4.96 20.5L6.76 18.71L5.34 17.29M12 6C8.69 6 6 8.69 6 12S8.69 18 12 18 18 15.31 18 12C18 8.68 15.31 6 12 6M20 13H23V11H20M17.24 18.71L19.04 20.5L20.45 19.09L18.66 17.29M20.45 5L19.04 3.6L17.24 5.39L18.66 6.81M13 1H11V4H13M6.76 5.39L4.96 3.6L3.55 5L5.34 6.81L6.76 5.39M1 13H4V11H1M13 20H11V23H13"></path>
			</svg>
		</h2>
	</header>
	<main id="content" class="content" role="main">
<article class="post">
    <span class="post-meta">
        <time datetime="2019-01-08">Jan 08, 2019</time>
        in <a href="https://blog.liang2.tw/category/bioinfo/">Bioinfo</a>
    </span>
    <h1 class="post-title">
        <a href="https://blog.liang2.tw/posts/2019/01/build-ensdb-from-local-mysql/" rel="bookmark" title="Permalink to Build EnsDb from a local Ensembl MySQL database">Build EnsDb from a local Ensembl MySQL database</a>
    </h1>
    <span class="post-meta">
                <a href="https://blog.liang2.tw/tag/en/">en</a>
                <a href="https://blog.liang2.tw/tag/r/">r</a>
                <a href="https://blog.liang2.tw/tag/bioconductor/">bioconductor</a>
                <a href="https://blog.liang2.tw/tag/sqlite/">sqlite</a>
                <a href="https://blog.liang2.tw/tag/ensembldb/">ensembldb</a>
    </span>
    <section class="post-content"><p>In some occasions, I need to access the older version of Ensembl human transcripts. For example, the mutation calls generated by the <a href="https://gdc.cancer.gov/">NCI&rsquo;s Genomic Data Common</a> pipeline are annotated by Ensembl v84. To programmatically query the Ensembl annotations, I use the EnsDb SQLite database created by <a href="https://bioconductor.org/packages/release/bioc/html/ensembldb.html">ensembldb</a>, which is a R package I enjoy using (see <a href="https://blog.liang2.tw/posts/2017/11/use-ensdb-database-in-python/">my previous post</a> for its usage).</p>
<p>The EnsDbs of the recent versions of Ensembl (v87+) are available on AnnotationHub. However, the older versions are not available, and they don&rsquo;t get updated when ensembldb introduces a new feature. For example, now newer EnsDbs include the transcript and gene ID version (<a href="https://github.com/jotsetung/ensembldb/issues/89">github issue</a>).</p>
<p>In my case, I need to build a EnsDb of Ensembl v84. <a href="https://bioconductor.org/packages/release/bioc/vignettes/ensembldb/inst/doc/ensembldb.html#10_getting_or_building_ensdb_databasespackages">The ensembldb&rsquo;s documentation</a> describes how to build one from the public Ensembl MySQL server. However, this method will take more than a day to complete. I started to look for other methods. After some trial and error, I managed to create my EnsDb fast by connecting to a local Ensembl database that I built. Surprisingly the setup wasn&rsquo;t difficult at all, and it only took about an hour to build the EnsDb.</p>
<p>Here are my notes of how to create the EnsDB from a local Ensembl MySQL database. I use macOS but the steps can be easily modified to work on other OSes.</p>
<div class="toc">
<ul>
<li><a href="#ensembl-vm">Ensembl VM</a></li>
<li><a href="#build-a-local-ensembl-v84-mysql-database">Build a local Ensembl v84 MySQL database</a></li>
<li><a href="#build-ensdb-from-the-local-mysql-database">Build EnsDB from the local MySQL database</a></li>
<li><a href="#remove-mysql">Remove MySQL</a></li>
</ul>
</div>
<h3 id="ensembl-vm">Ensembl VM</h3>
<p>To create a EnsDB from a Ensembl MySQL database, we need to the Ensembl Perl APIs. And the easiest setup is by a <a href="http://www.ensembl.org/info/data/virtual_machine.html">Ensembl virtual machine</a>. We just need to import the VM image using VirtualBox and install the ensembldb R package inside the      VM, then it is ready to build the EnsDb. I recommend the VM to have more memory than the default 1GB since a larger memory helps build the R packages and EnsDb.</p>
<h3 id="build-a-local-ensembl-v84-mysql-database">Build a local Ensembl v84 MySQL database</h3>
<p>Ensembl provides <a href="https://www.ensembl.org/info/docs/webcode/mirror/install/ensembl-data.html">the MySQL database dump</a> to allow easy import of their data of any version. Assuming the working directory is <code>~/Documents/Ensembl_MySQL_mirror/</code>, we first copy the database dump by:</p>
<div class="highlight"><pre><span></span><code class="language-bash"><span class="nb">cd</span> ~/Documents/Ensembl_MySQL_mirror

<span class="c1"># Download the db dump</span>
rsync -a rsync://ftp.ensembl.org/ensembl/pub/release-84/mysql/homo_sapiens_core_84_38 .

<span class="c1"># MySQL doesn&#39;t accept compressed db dump files so we decompress them</span>
gunzip *.txt.gz
</code></pre></div>

<p>While downloading the data, we also need to install the MySQL server. I install <a href="http://www.ensembl.org/info/data/mysql.html">the same or similar version of MySQL</a> Ensembl is currently using, which is 5.6 at the time of writing. On macOS, Homebrew can specify the version of MySQL to be installed:</p>
<div class="highlight"><pre><span></span><code class="language-bash">brew install mysql@5.6
<span class="c1"># And launch the MySQL server</span>
/usr/local/opt/mysql@5.6/bin/mysql.server start
</code></pre></div>

<p>First we create a database whose name matches the Ensembl version (v84):</p>
<div class="highlight"><pre><span></span><code class="language-sql"><span class="k">CREATE</span><span class="w"> </span><span class="k">DATABASE</span><span class="w"> </span><span class="n">homo_sapiens_core_84_38</span><span class="p">;</span><span class="w"></span>
</code></pre></div>

<p>Then we load the table schema and Ensembl data:</p>
<div class="highlight"><pre><span></span><code class="language-bash">/usr/local/opt/mysql@5.6/bin/mysql -u root <span class="se">\</span>
    homo_sapiens_core_84_38 &lt; homo_sapiens_core_84_38.sql

/usr/local/opt/mysql@5.6/bin/mysqlimport <span class="se">\</span>
    -u root <span class="se">\</span>
    --fields-terminated-by<span class="o">=</span><span class="s1">&#39;\t&#39;</span> --fields_escaped_by<span class="o">=</span><span class="se">\\</span>  <span class="se">\</span>
    homo_sapiens_core_84_38 -L *.txt
</code></pre></div>

<p>Finally, we modify the MySQL config at <code>/usr/local/etc/my.cnf</code> to accept remote database connection, so our VM can access the database on the host machine. I don&rsquo;t use MySQL for anything else, so I simply let MySQL binds to all the possible IP addresses my machine has:</p>
<div class="highlight"><pre><span></span><code class="language-text">[mysqld]
bind-address = *
</code></pre></div>

<p>Note that this is not a secure configuration. To be secure, there should be a designated MySQL user with limited permission and a stricter connection setting. Restart MySQL to load the new config:</p>
<div class="highlight"><pre><span></span><code class="language-text">/usr/local/opt/mysql@5.6/bin/mysql.server restart
</code></pre></div>

<p>Write down an (local) IP address of our host machine.</p>
<h3 id="build-ensdb-from-the-local-mysql-database">Build EnsDB from the local MySQL database</h3>
<p>Now we can come back to the vm and build the EnsDb v84. Run the following R script to build the EnsDb:</p>
<div class="highlight"><pre><span></span><code class="language-splus"><span class="nf">library</span><span class="p">(</span><span class="n">ensembldb</span><span class="p">)</span>
<span class="nf">fetchTablesFromEnsembl</span><span class="p">(</span>
    <span class="m">84</span><span class="p">,</span> <span class="n">species</span> <span class="o">=</span> <span class="s">&quot;human&quot;</span><span class="p">,</span>
    <span class="n">user</span> <span class="o">=</span> <span class="s">&#39;root&#39;</span><span class="p">,</span> <span class="n">host</span> <span class="o">=</span> <span class="s">&#39;&lt;our host IP&gt;&#39;</span><span class="p">,</span> <span class="n">port</span> <span class="o">=</span> <span class="m">3306</span>
<span class="p">)</span>
<span class="n">DBFile</span> <span class="o">&lt;-</span> <span class="nf">makeEnsemblSQLiteFromTables</span><span class="p">()</span>
</code></pre></div>

<p>The EnsDb SQLite database will be availabe under the working directory. We can test the new EnsDb by:</p>
<div class="highlight"><pre><span></span><code class="language-splus"><span class="n">edb</span> <span class="o">&lt;-</span> <span class="nf">EnsDb</span><span class="p">(</span><span class="n">DBFile</span><span class="p">)</span>
</code></pre></div>

<h3 id="remove-mysql">Remove MySQL</h3>
<p>If there is no other need of MySQL, we can uninstall it and remove all its databases by:</p>
<div class="highlight"><pre><span></span><code class="language-text">brew remove mysql@5.6
rm -rf /usr/local/var/mysql
</code></pre></div></section>
    <footer class="post-footer">
	<section
		class="utterances"
		data-repo="ccwang002/ccwang002.github.io"
		data-issue-term="pathname"
		data-label="Comment">
	</section>
	</footer>
</article>
	</main>
	<footer class="site-footer">
		<a class="subscribe"
			href="https://blog.liang2.tw/feeds/all.atom.xml"
			data-tooltip="Atom feed"
		>
			<svg viewBox="0 0 24 24" width="24" height="24" alt="Atom feed">
				<circle cx="6.18" cy="17.82" r="2.18"/>
				<path d="m4 10.1v2.83c3.9 0 7.07 3.17 7.07 7.07h2.83c0-5.47-4.43-9.9-9.9-9.9zm0-5.66v2.83c7.03 0 12.73 5.7 12.73 12.73h2.83c0-8.59-6.97-15.56-15.56-15.56z"/>
			</svg>
		</a>
		<div class="inner">
			<section>
				Browse articles by
				<a href="https://blog.liang2.tw/tags.html">Tags</a> |
				<a href="https://blog.liang2.tw/categories.html">Categories</a>
			</section>
			<section class="contact">
				Contact me by
				<a href="https://twitter.com/lbwang2">Twitter</a> |
				<a href="https://www.facebook.com/lbwang.2">Facebook</a> |
				<a href="https://github.com/ccwang002">GitHub</a> |
				<a href="mailto:me+blog@liang2.tw">Email</a> |
				<a href="https://www.linkedin.com/in/liangbowang/">LinkedIn</a>
			</section>
			<section class="copyright">
				This work is licensed under <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.
			</section>
			<section class="powered-by">
				Built using <a href="https://getpelican.com/">Pelican</a>.
				Powered by the <a href="https://github.com/kura/hauntr">Hauntr</a> theme.
			</section>
		</div>
	</footer>
	<script src="https://blog.liang2.tw/theme/scripts/theme-toggle.js"></script>
</body>
</html>