<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="utf-8" />
	<meta http-equiv="X-UA-Compatible" content="IE=edge, chrome=1">
	<meta http-equiv="content-type" content="text/html; charset=UTF-8;charset=utf-8">
	<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="author" content="Liang-Bo Wang" />
<meta name="description" content="I often received data that was annotated by an unknown Ensembl release/version. It could be the Ensembl IDs in a gene expression matrix, a VEP annotated MAF file, or even a customized GTF. The documentation of those files wasnâ€™t always clear about the annotation in use. However, it â€¦" />
<meta name="keywords" content="en">
	<link rel="icon" href="data:image/svg+xml,<svg xmlns=%22http://www.w3.org/2000/svg%22 viewBox=%220 0 100 100%22><text y=%22.95em%22 font-size=%2295%22>ðŸ“–</text></svg>">
	<title>Identify the Ensembl release from versioned IDs</title>

	<!-- og -->
<meta property="og:locale" content="en_US"/>
<meta property="og:site_name" content="Liang-Bo Wang's Blog"/>
<meta property="og:title" content="Identify the Ensembl release from versioned IDs"/>
<meta property="og:description" content="I often received data that was annotated by an unknown Ensembl release/version. It could be the Ensembl IDs in a gene expression matrix, a VEP annotated MAF file, or even a customized GTF. The documentation of those files wasnâ€™t always clear about the annotation in use. However, it â€¦"/>
<meta property="og:locale" content=""/>
<meta property="og:url" content="https://blog.liang2.tw/posts/2020/06/identify-ensembl-release-ver/"/>
<meta property="og:type" content="article"/>
<meta property="article:published_time" content="2020-06-08 00:00:00-05:00"/>
<meta property="article:modified_time" content="2024-09-08 17:21:08.419717-05:00"/>
<meta property="article:author" content="https://blog.liang2.tw/author/liang-bo-wang.html">
<meta property="article:section" content="Bioinfo"/>
<meta property="article:tag" content="en"/>
<meta name="twitter:card" content="summary">
<meta name="twitter:site" content="@ccwang002">
<meta name="twitter:title" content="Identify the Ensembl release from versioned IDs">
<meta name="twitter:description" content="I often received data that was annotated by an unknown Ensembl release/version. It could be the Ensembl IDs in a gene expression matrix, a VEP annotated MAF file, or even a customized GTF. The documentation of those files wasnâ€™t always clear about the annotation in use. However, it â€¦">

	<!-- Feeds -->
	<link href="https://blog.liang2.tw/feeds/all.atom.xml"
		type="application/atom+xml" rel="alternate" title="Liang-Bo Wang's Blog" />

	<link rel="stylesheet" type="text/css" href="https://blog.liang2.tw/theme/styles/styles.css" />
	<link rel="stylesheet" href="https://blog.liang2.tw/theme/styles/pygments/light.css" type="text/css" media="(prefers-color-scheme: light)" id="pyg-light" />
    <link rel="stylesheet" href="https://blog.liang2.tw/theme/styles/pygments/dark-monokai.css" type="text/css" media="(prefers-color-scheme: dark)" id="pyg-dark" />
	<link rel="stylesheet" type="text/css" href="https://blog.liang2.tw/theme/katex-0.16.2/katex.min.css">
</head>

<body class="post-template no-js">
	<script>document.body.classList.remove('no-js');</script>
<svg
	xmlns="http://www.w3.org/2000/svg"
	style="display: none;"
>
	<symbol id="svg-sun-half" viewBox="0 0 24 24" pointer-events="all">
		<title>Following system color scheme</title>
		<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none"
			stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round">
			<circle cx="12" cy="12" r="9"></circle>
			<path d="M12 3v18m0-12l4.65-4.65M12 14.3l7.37-7.37M12 19.6l8.85-8.85"></path>
		</svg>
	</symbol>
	<symbol id="svg-moon" viewBox="0 0 24 24" pointer-events="all">
		<title>Selected dark color scheme</title>
		<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none"
			stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round">
			<path stroke="none" d="M0 0h24v24H0z" fill="none"></path>
			<path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z"></path>
		</svg>
	</symbol>
	<symbol id="svg-sun" viewBox="0 0 24 24" pointer-events="all">
		<title>Selected light color scheme</title>
		<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none"
			stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round">
			<circle cx="12" cy="12" r="5"></circle>
			<line x1="12" y1="1" x2="12" y2="3"></line>
			<line x1="12" y1="21" x2="12" y2="23"></line>
			<line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
			<line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
			<line x1="1" y1="12" x2="3" y2="12"></line>
			<line x1="21" y1="12" x2="23" y2="12"></line>
			<line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
			<line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
		</svg>
	</symbol>
</svg>

<script>
	document.documentElement.dataset.color_scheme = localStorage.getItem("color_scheme") || "auto"
</script>
	<header class="site-head">
		<h1 class="blog-title"><a href="https://blog.liang2.tw">Liang-Bo Wang's Blog</a></h1>
		<h2 class="blog-description">
			<a href="https://blog.liang2.tw/about-me/">About</a> |
			<a href="https://blog.liang2.tw/talks/">Talks</a> |
			<a href="https://blog.liang2.tw/archives.html">Archives</a> |
			<button id="color-scheme-cycler" onClick="setColorScheme(nextColorScheme())">
                <svg aria-hidden="true" class="color-scheme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
                <svg aria-hidden="true" class="color-scheme-icon-when-dark"><use href="#svg-moon"></use></svg>
                <svg aria-hidden="true" class="color-scheme-icon-when-light"><use href="#svg-sun"></use></svg>
                <span class="sr-only">Toggle light / dark / auto color theme</span>
            </button>
		</h2>
	</header>
	<main id="content" class="content" role="main">
<article class="post">
    <span class="post-meta">
        <time datetime="2020-06-08">Jun 08, 2020</time>
        in <a href="https://blog.liang2.tw/category/bioinfo/">Bioinfo</a>
    </span>
    <h1 class="post-title">
        <a href="https://blog.liang2.tw/posts/2020/06/identify-ensembl-release-ver/" rel="bookmark" title="Permalink to Identify the Ensembl release from versioned IDs">Identify the Ensembl release from versioned IDs</a>
    </h1>
    <span class="post-meta">
                <a href="https://blog.liang2.tw/tag/en/">en</a>
    </span>
    <section class="post-content"><p>I often received data that was annotated by an unknown Ensembl release/version.</p>
<p>It could be the Ensembl IDs in a gene expression matrix, a VEP annotated MAF file, or even a customized GTF. The documentation of those files wasn&rsquo;t always clear about the annotation in use. However, it&rsquo;s sometimes necessary to know the exact Ensembl release. Say, I want to reproduce the result, or to pass the output to the extended downstream workflow. While the tiny difference between adjacent releases is only annoying, <a href="https://betatark.ensembl.org/web/statistics/">thousands of changes per release</a> add up quickly to be obvious inconsistency when using releases across different years.</p>
<p>It&rsquo;s possible to pinpoint the Ensembl release using the ID versions. For example, ENSG00000119772.15 (DNMT3A) only existed in Ensembl releases 79 and 80; ENST00000275493.7 (EGFR) remains alive since release 96. By checking the ID history on <a href="https://ensembl.org">https://ensembl.org</a>, I can identify the possible Ensembl releases my data uses. A handy URL shortcut to accompany the investigation is <code>ensembl.org/id/&lt;ensembl_id&gt;</code>, which can redirect to the different page tabs depends on the ID types (e.g., ENSG to genes and ENST to transcript).</p>
<figure>
  <img src="https://blog.liang2.tw/posts/2020/06/identify-ensembl-release-ver/pics/ensembl_r100_dnmt3A_id_history.png">
  <figcaption>ID history of ENSG00000119772 (DNMT3A) (<a href="https://www.ensembl.org/Homo_sapiens/Gene/Idhistory?g=ENSG00000119772">source</a>)</figcaption>
</figure>

<p>With <a href="https://betatark.ensembl.org/">Ensembl Tark</a>, I wrote <a href="https://gist.github.com/ccwang002/829a5420a47adfb3be597ed3ea8a0a29">a Python script</a> to automate the query. Given a list of versioned Ensembl IDs<sup id="fnref:versioned-id"><a class="footnote-ref" href="#fn:versioned-id">1</a></sup>, the script will identify the possible Ensembl releases. The IDs can be genes, transcripts, or proteins. Based on my testing, the script can identify Ensembl release range with less than 30 IDs.</p>
<p>Here is an example list of IDs:</p>
<div class="highlight"><pre><span></span><code>ENST00000426449.4
ENST00000434817.4
ENSP00000222254.6
ENSP00000477864.1
ENSG00000170266.14
ENSG00000238009.5
ENSG00000173705.7
</code></pre></div>

<p>And the output of the script:</p>
<div class="highlight"><pre><span></span><code><span class="gp">$ </span>python<span class="w"> </span>check_possible_ensembl_releases.py<span class="w"> </span>ensembl_ids.list
<span class="go">[2020-06-09 13:00:48][INFO   ] Querying for 7 Ensembl IDs</span>
<span class="go">[2020-06-09 13:00:49][INFO   ] Only Ensembl releases 75â€“99 are supported by Ensembl Tark. IDs outside the range may not be identified.</span>
<span class="go">ENSP00000477864.1 in Ensembl releases 76â€“99</span>
<span class="go">ENSG00000173705.7 in Ensembl releases 79â€“80</span>
<span class="go">ENSG00000238009.5 in Ensembl releases 79â€“80</span>
<span class="go">ENST00000434817.4 in Ensembl releases 79â€“80</span>
<span class="go">ENSP00000222254.6 in Ensembl releases 75â€“99</span>
<span class="go">ENSG00000170266.14 in Ensembl releases 79â€“80</span>
<span class="go">ENST00000426449.4 in Ensembl releases 79â€“80</span>
<span class="go">Possible Ensembl releases are: 79, 80</span>
</code></pre></div>

<p>In this case, Ensembl releases 79 and 80 have the same gene model.</p>
<p>The script basically uses Tark&rsquo;s REST APIs to get the release range for each ID, and find the intersection of all the ranges. It runs on Python 3.8 and uses aiohttp 3.6 for concurrent API calls. Tark currently has records from release 75 (2014) to 99 (2020), so this approach will fail if the IDs are too old (or too new, but I think it will include the latest r100 soon). I limited the maximal concurrent calls â‰¤ 5 so I don&rsquo;t overwhelm the Tark service.</p>
<p><a href="https://betatark.ensembl.org/">Tark</a> is a great website/service that compares the transcripts between different versions or even different sources (Ensembl vs RefSeq)! It can tell you what exon or UTR was changed, something quick tricky to set up because one has to import databases for every Ensembl release and even RefSeq releases. Tark is currently in beta, but I hope it can be stable and remained updated.</p>
<p>And this is just another point that it&rsquo;s usually a good idea to include versioned IDs in the data.</p>
<p>Now I can go back digging the history of my data :)</p>
<div class="footnote">
<hr>
<ol>
<li id="fn:versioned-id">
<p>It&rsquo;s probably possible to apply the same approach without Ensembl ID version (e.g., ENSG00000119772), but one might need a lot of them because IDs are much more stable.&#160;<a class="footnote-backref" href="#fnref:versioned-id" title="Jump back to footnote 1 in the text">&#8617;</a></p>
</li>
</ol>
</div></section>
    <footer class="post-footer">
	<section
		class="utterances"
		data-repo="ccwang002/ccwang002.github.io"
		data-issue-term="pathname"
		data-label="Comment">
	</section>
	</footer>
</article>
	</main>
	<footer class="site-footer">
		<a class="subscribe"
			href="https://blog.liang2.tw/feeds/all.atom.xml"
			aria-label="Subscribe the Atom feed"
			data-tooltip="Subscribe"
		>
			<svg viewBox="0 0 24 24" width="24" height="24">
				<title>Atom feed</title>
				<circle cx="6.18" cy="17.82" r="2.18"/>
				<path d="m4 10.1v2.83c3.9 0 7.07 3.17 7.07 7.07h2.83c0-5.47-4.43-9.9-9.9-9.9zm0-5.66v2.83c7.03 0 12.73 5.7 12.73 12.73h2.83c0-8.59-6.97-15.56-15.56-15.56z"/>
			</svg>
		</a>
		<div class="inner">
			<section>
				Browse articles by
				<a href="https://blog.liang2.tw/tags.html">Tags</a> |
				<a href="https://blog.liang2.tw/categories.html">Categories</a>
			</section>
			<section class="contact">
				Contact me by
				<a href="https://twitter.com/lbwang2">Twitter</a> |
				<a href="https://www.facebook.com/lbwang.2">Facebook</a> |
				<a href="https://github.com/ccwang002">GitHub</a> |
				<a href="mailto:me+blog@liang2.tw">Email</a> |
				<a href="https://www.linkedin.com/in/liangbowang/">LinkedIn</a>
			</section>
			<section class="copyright">
				This work is licensed under <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.
			</section>
			<section class="powered-by">
				Built using <a href="https://getpelican.com/">Pelican</a>.
				Powered by the <a href="https://github.com/kura/hauntr">Hauntr</a> theme.
			</section>
		</div>
	</footer>
	<script src="https://blog.liang2.tw/theme/scripts/color_scheme.js"></script>
</body>
</html>