<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="utf-8" />
	<meta http-equiv="X-UA-Compatible" content="IE=edge, chrome=1">
	<meta http-equiv="content-type" content="text/html; charset=UTF-8;charset=utf-8">
	<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="author" content="Liang-Bo Wang" />
<meta name="description" content="Notes of fundamental tools and learning resources for handling genomic data in R with Bioconductor." />
<meta name="keywords" content="en, r, bioconductor">
	<link rel="icon" href="data:image/svg+xml,<svg xmlns=%22http://www.w3.org/2000/svg%22 viewBox=%220 0 100 100%22><text y=%22.95em%22 font-size=%2295%22>ðŸ“–</text></svg>">
	<title>Overview of Genomic Data Processing in Bioconductor</title>

	<!-- og -->
<meta property="og:locale" content="en_US"/>
<meta property="og:site_name" content="Liang-Bo Wang's Blog"/>
<meta property="og:title" content="Overview of Genomic Data Processing in Bioconductor"/>
<meta property="og:description" content="Notes of fundamental tools and learning resources for handling genomic data in R with Bioconductor."/>
<meta property="og:locale" content=""/>
<meta property="og:url" content="https://blog.liang2.tw/posts/2015/12/biocondutor-genomic-data/"/>
<meta property="og:type" content="article"/>
<meta property="article:published_time" content="2015-12-29 20:28:00-06:00"/>
<meta property="article:modified_time" content="2022-02-20 20:07:45.273382-06:00"/>
<meta property="article:author" content="https://blog.liang2.tw/author/liang-bo-wang.html">
<meta property="article:section" content="Bioinfo"/>
<meta property="article:tag" content="en"/>
<meta property="article:tag" content="r"/>
<meta property="article:tag" content="bioconductor"/>
<meta name="twitter:card" content="summary">
<meta name="twitter:site" content="@ccwang002">
<meta name="twitter:title" content="Overview of Genomic Data Processing in Bioconductor">
<meta name="twitter:description" content="Notes of fundamental tools and learning resources for handling genomic data in R with Bioconductor.">

	<!-- Feeds -->
	<link href="https://blog.liang2.tw/feeds/all.atom.xml"
		type="application/atom+xml" rel="alternate" title="Liang-Bo Wang's Blog" />

	<link rel="stylesheet" type="text/css" href="https://blog.liang2.tw/theme/styles/styles.css" />
	<link rel="stylesheet" href="https://blog.liang2.tw/theme/styles/pygments/light.css" type="text/css" media="(prefers-color-scheme: light)" id="pyg-light" />
    <link rel="stylesheet" href="https://blog.liang2.tw/theme/styles/pygments/dark-monokai.css" type="text/css" media="(prefers-color-scheme: dark)" id="pyg-dark" />
	<link rel="stylesheet" type="text/css" href="https://blog.liang2.tw/theme/katex-0.16.2/katex.min.css">
</head>

<body class="post-template no-js">
	<script>document.body.classList.remove('no-js');</script>
<svg
	xmlns="http://www.w3.org/2000/svg"
	style="display: none;"
>
	<symbol id="svg-sun-half" viewBox="0 0 24 24" pointer-events="all">
		<title>Following system color scheme</title>
		<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none"
			stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round">
			<circle cx="12" cy="12" r="9"></circle>
			<path d="M12 3v18m0-12l4.65-4.65M12 14.3l7.37-7.37M12 19.6l8.85-8.85"></path>
		</svg>
	</symbol>
	<symbol id="svg-moon" viewBox="0 0 24 24" pointer-events="all">
		<title>Selected dark color scheme</title>
		<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none"
			stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round">
			<path stroke="none" d="M0 0h24v24H0z" fill="none"></path>
			<path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z"></path>
		</svg>
	</symbol>
	<symbol id="svg-sun" viewBox="0 0 24 24" pointer-events="all">
		<title>Selected light color scheme</title>
		<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none"
			stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round">
			<circle cx="12" cy="12" r="5"></circle>
			<line x1="12" y1="1" x2="12" y2="3"></line>
			<line x1="12" y1="21" x2="12" y2="23"></line>
			<line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
			<line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
			<line x1="1" y1="12" x2="3" y2="12"></line>
			<line x1="21" y1="12" x2="23" y2="12"></line>
			<line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
			<line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
		</svg>
	</symbol>
</svg>

<script>
	document.documentElement.dataset.color_scheme = localStorage.getItem("color_scheme") || "auto"
</script>
	<header class="site-head">
		<h1 class="blog-title"><a href="https://blog.liang2.tw">Liang-Bo Wang's Blog</a></h1>
		<h2 class="blog-description">
			<a href="https://blog.liang2.tw/about-me/">About</a> |
			<a href="https://blog.liang2.tw/talks/">Talks</a> |
			<a href="https://blog.liang2.tw/archives.html">Archives</a> |
			<button id="color-scheme-cycler" onClick="setColorScheme(nextColorScheme())">
                <svg aria-hidden="true" class="color-scheme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
                <svg aria-hidden="true" class="color-scheme-icon-when-dark"><use href="#svg-moon"></use></svg>
                <svg aria-hidden="true" class="color-scheme-icon-when-light"><use href="#svg-sun"></use></svg>
                <span class="sr-only">Toggle light / dark / auto color theme</span>
            </button>
		</h2>
	</header>
	<main id="content" class="content" role="main">
<article class="post">
    <span class="post-meta">
        <time datetime="2015-12-29">Dec 29, 2015</time>
        in <a href="https://blog.liang2.tw/category/bioinfo/">Bioinfo</a>
    </span>
    <h1 class="post-title">
        <a href="https://blog.liang2.tw/posts/2015/12/biocondutor-genomic-data/" rel="bookmark" title="Permalink to Overview of Genomic Data Processing in Bioconductor">Overview of Genomic Data Processing in Bioconductor</a>
    </h1>
    <span class="post-meta">
                <a href="https://blog.liang2.tw/tag/en/">en</a>
                <a href="https://blog.liang2.tw/tag/r/">r</a>
                <a href="https://blog.liang2.tw/tag/bioconductor/">bioconductor</a>
    </span>
    <section class="post-content"><p>Sorry for the late update. In the past two months, I finished my Ph.D. applications (hope to hear good news in the next two months) and was busy preparing the PyCon Taiwan 2016. Also, a year-long website development finally came to the end.</p>
<p>Now most things are set so I can back to writing my blog.</p>
<p>Since September, there accumulates at least 5 drafts and I don&rsquo;t know when I can finish them, so I think I have to change my writing strategy. I will first publish things as soon as information collection is done, and deeper reviews will be given in the following posts. Right now I will focus on Bioconductor (and general Bioinformatics topics) and Django.</p>
<div class="toc">
<ul>
<li><a href="#bioconductor">Bioconductor</a><ul>
<li><a href="#annotation-and-genome-reference">Annotation and Genome Reference</a></li>
<li><a href="#experiment-data-storage">Experiment Data Storage</a></li>
<li><a href="#operations-on-genome">Operations on Genome</a></li>
<li><a href="#genomic-data-visualization">Genomic data visualization</a></li>
</ul>
</li>
<li><a href="#summary">Summary</a></li>
</ul>
</div>
<h2 id="bioconductor">Bioconductor</h2>
<p><a href="https://www.bioconductor.org/">Bioconductor</a> is indeed a rich resources for R both in terms of data and tools. And I found I have yet spent time seriously understanding the whole ecosystem, which I believe can drastically lighten the loading of daily analysis.</p>
<p>Bioconductor&rsquo;s website is informative. If you are familar with R, you should already know that in order to understand the usage of a package, one of the best way is to read its vignettes. Packages on Bioconductor generally have vignettes, which is really helpful and the website makes them accessible. On top of that, they have <a href="https://www.bioconductor.org/help/course-materials/">Courses &amp; Conferences</a> and <a href="https://www.bioconductor.org/help/workflows/">Workflows</a>. The former section collects all conference materials in the past few years, which contains package hands-on, analysis tutorial, and R advanced topics. It&rsquo;s a hidden gem to me since I have already found numerous materials worth reading only after a glance over it. The latter one should be well-known. It gives examples of typical analysis workflows.</p>
<p>I&rsquo;m interested in the following topics in Biocondutor:</p>
<ul>
<li>Annotation and genome reference (OrgDb, TxDb, OrganismDb, BSgenome)</li>
<li>Experiment data storage (ExpressionSets)</li>
<li>Operations on genome (GenomicRanges)</li>
<li>Genomic data visualization (Gviz, ggbio)</li>
</ul>
<p>Keywords in Biocondutors for each topic are attached in the parens, mostly being the package name. For each topic, I&rsquo;ll put the related resources I collected in the following sections.</p>
<p>Before the listing, I found <a href="http://genomicsclass.github.io/book/">PH525x series</a> maintained by Rafael Irizarry and Michael Love from Harvard serves as a comprehensive entry point for almost every related topic. The site is the accompanied resources for their edX classes. Both of them worth taking a look.</p>
<h3 id="annotation-and-genome-reference">Annotation and Genome Reference</h3>
<ul>
<li>
<p><a href="http://genomicsclass.github.io/book/pages/annoPhen.html">Annotating phenotypes and molecular function</a> from <a href="http://genomicsclass.github.io/book/">PH525x series</a> gives a good overview and a taste of the powerful ecosystem Bioconductor provides.</p>
</li>
<li>
<p><a href="https://www.bioconductor.org/help/course-materials/2015/BioC2015/Annotation_Resources.html">Annotation Resources</a> from <a href="https://www.bioconductor.org/help/course-materials/2015/BioC2015/">BioC 2015</a> gives more extensive introduction about all available types of references from genome sequences to transcriptome and gene info.</p>
</li>
</ul>
<p>For example, human comes with</p>
<ul>
<li><a href="http://bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html">org.Hs.eg.db</a></li>
<li><a href="http://bioconductor.org/packages/release/data/annotation/html/TxDb.Hsapiens.UCSC.hg38.knownGene.html">TxDb.Hsapiens.UCSC.hg38.knownGene</a></li>
<li><a href="http://bioconductor.org/packages/release/data/annotation/html/Homo.sapiens.html">Homo.sapiens</a></li>
<li><a href="http://bioconductor.org/packages/release/data/annotation/html/BSgenome.Hsapiens.UCSC.hg38.html">BSgenome.Hsapiens.UCSC.hg38</a></li>
</ul>
<h3 id="experiment-data-storage">Experiment Data Storage</h3>
<p>ExpressionSet helps store the expression experiment data, which one can combine   expression values and phenotypes of the same sample. Additionally the experiment data (like descriptions of GEO dataset) can be attached as well.</p>
<ul>
<li>
<p><a href="http://genomicsclass.github.io/book/pages/eset.html">The ExpressionSet container</a> from <a href="http://genomicsclass.github.io/book/">PH525x series</a> gives an intro. It should be sufficient enough to use ExpressionSet in daily work.</p>
</li>
<li>
<p><a href="https://www.bioconductor.org/packages/release/bioc/vignettes/Biobase/inst/doc/ExpressionSetIntroduction.pdf">The ExpressionSet Introduction</a> from its package <a href="https://www.bioconductor.org/packages/release/bioc/html/Biobase.html">Biobase</a>â€™s vignette gives detailed explanation.</p>
</li>
</ul>
<h3 id="operations-on-genome">Operations on Genome</h3>
<p>I haven&rsquo;t gone into the details, but operations about genomic ranges are often tricky and more importantly, badly optimized.</p>
<ul>
<li>
<p><a href="http://genomicsclass.github.io/book/pages/iranges_granges.html">IRanges and GRanges</a> and <a href="http://genomicsclass.github.io/book/pages/operateGRanges.html">GRanges operations</a> from <a href="http://genomicsclass.github.io/book/">PH525x series</a> give the overview of using the package <a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges</a>.</p>
</li>
<li>
<p><a href="https://www.bioconductor.org/packages/release/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.pdf">An Introduction to Genomic Ranges Classes</a>, a <a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges</a> vignette, gives a detailed view.</p>
</li>
<li>
<p>Also, their paper, <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003118">&ldquo;Software for Computing and Annotating Genomic Ranges&rdquo;, <em>PLOS One</em></a> should be another overview source of the package.</p>
</li>
<li>
<p><a href="https://cran.r-project.org/web/packages/data.table/index.html">data.table</a>&lsquo;s <code>foverlap</code> function worth the comparison, since I already use it and I know it is <a href="https://github.com/Rdatatable/data.table/wiki/talks/EARL2014_OverlapRangeJoin_Arun.pdf">blazingly fast</a>. <code>foverlap</code> handles the overlapping of integer ranges so it can be applied to genomic operation. Its code is quite complex so its mechanism is still a myth to me. I&rsquo;d like to see its comparison with using database like SQLite.</p>
</li>
</ul>
<h3 id="genomic-data-visualization">Genomic data visualization</h3>
<p>Basically I can find two packages:</p>
<ul>
<li><a href="https://bioconductor.org/packages/release/bioc/html/Gviz.html">Gviz</a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/ggbio.html">ggbio</a></li>
</ul>
<p>Don&rsquo;t know their difference yet. Both of them can produce well-done figures. But I think I have some experience with ggbio, which was a bit tricky to use. So for now I will go for Gviz.</p>
<ul>
<li><a href="https://www.bioconductor.org/help/course-materials/2012/BiocEurope2012/GvizEuropeanBioc2012.pdf">Visualizing genomic features with the Gviz package</a> given at Bioc Europe 2012 has a decent introduction about Gviz.</li>
<li><a href="https://bioconductor.org/packages/release/bioc/vignettes/Gviz/inst/doc/Gviz.pdf">The Gviz User Guide</a> looks very comprehensive, which also cover usage with expression and alignment results.</li>
</ul>
<h2 id="summary">Summary</h2>
<p>These resources should be enough for weeks of trying. It&rsquo;s excited to find so many useful tools.</p>
<p>So, good luck to me for my Ph.D. application, PyCon Taiwan 2016, and a shorter blog posting frequency.</p></section>
    <footer class="post-footer">
	<section
		class="utterances"
		data-repo="ccwang002/ccwang002.github.io"
		data-issue-term="pathname"
		data-label="Comment">
	</section>
	</footer>
</article>
	</main>
	<footer class="site-footer">
		<a class="subscribe"
			href="https://blog.liang2.tw/feeds/all.atom.xml"
			aria-label="Subscribe the Atom feed"
			data-tooltip="Subscribe"
		>
			<svg viewBox="0 0 24 24" width="24" height="24">
				<title>Atom feed</title>
				<circle cx="6.18" cy="17.82" r="2.18"/>
				<path d="m4 10.1v2.83c3.9 0 7.07 3.17 7.07 7.07h2.83c0-5.47-4.43-9.9-9.9-9.9zm0-5.66v2.83c7.03 0 12.73 5.7 12.73 12.73h2.83c0-8.59-6.97-15.56-15.56-15.56z"/>
			</svg>
		</a>
		<div class="inner">
			<section>
				Browse articles by
				<a href="https://blog.liang2.tw/tags.html">Tags</a> |
				<a href="https://blog.liang2.tw/categories.html">Categories</a>
			</section>
			<section class="contact">
				Contact me by
				<a href="https://twitter.com/lbwang2">Twitter</a> |
				<a href="https://www.facebook.com/lbwang.2">Facebook</a> |
				<a href="https://github.com/ccwang002">GitHub</a> |
				<a href="mailto:me+blog@liang2.tw">Email</a> |
				<a href="https://www.linkedin.com/in/liangbowang/">LinkedIn</a>
			</section>
			<section class="copyright">
				This work is licensed under <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.
			</section>
			<section class="powered-by">
				Built using <a href="https://getpelican.com/">Pelican</a>.
				Powered by the <a href="https://github.com/kura/hauntr">Hauntr</a> theme.
			</section>
		</div>
	</footer>
	<script src="https://blog.liang2.tw/theme/scripts/color_scheme.js"></script>
</body>
</html>